Add RFC for Grid Catalog (#14)
hyperledger/grid-rfcs: 0000-catalog.md

exipire the catalog. Also an optional **Name** for the catalog. 

A catalog, like a product, can also have one or more **properties**.  Properties
are described in the Grid Primitives RFC.  A *property namespace* contains
multiple *property schemas*.  A property schema associates a name (such as
“length”) with a data type (such as integer).  

## Transactions

Catalogs are managed by submitting transactions to Hyperledger Grid, which will
process them with the Grid Catalog smart contract. The following transactions
are supported:

**Catalog actions:**
* CatalogCreate - create a catalog and store it in state
* CatalogDelete - removes a catalog from state
* CatalogReplicate - creates another copy of a catalog that already exists in state

**Catalog operations:**
* AddProductsToCatalog - adds a list of products specified by their product_id to the catalog 
* RemoveProductsFromCatalog - removes a product or products from a catalog
* ActivateProduct - toggles the "status" of a product to an active state
* DeactivateProduct - toggles the "status" of a product to an inactive state
* DiscontinueProduct - removes a product from all catalogs it's in

> 

Reviewer: Given that the bulk of these operations are 'about products', or maybe more specifically 'about catalog products', it may make sense to perform the operations on a 'product-like object' instead of a 'catalog-like object'. If there were a location in state referenced by the product identifier that contained a collection of catalog identifiers (and associated 'product catalog attributes'), all the 'catalog operations' listed here could be performed against that structure. This would be separate from the product records, and separate from the catalog record (which would define the catalog identifier and other attributes). This model may make more sense if the cardinality of products to catalogs is materially higher than catalogs to products. Drawbacks are that it makes it difficult to get a list of the products in a catalog without a reporting database. Depending on what we think the extents of 'number of products in a catalog' and/or 'number of catalogs a product is in' are, we may also want to consider other storage models for the collection, including some kind of paginating structure similar to the sensor record storage in track and trace.

Reviewer: Also, depending on the sizes expected, storing something like a space-efficient Bloom Filter on either the Catalog or ProductCatalogs records as a maintained existence map may be worthwhile.

Author: > Given that the bulk of these operations are 'about products', or maybe more specifically 'about catalog products', it may make sense to perform the operations on a 'product-like object' instead of a 'catalog-like object'. If there were a location in state referenced by the product identifier that contained a collection of catalog identifiers (and associated 'product catalog attributes'), all the 'catalog operations' listed here could be performed against that structure. This would be separate from the product records, and separate from the catalog record (which would define the catalog identifier and other attributes). This model may make more sense if the cardinality of products to catalogs is materially higher than catalogs to products. Drawbacks are that it makes it difficult to get a list of the products in a catalog without a reporting database. Depending on what we think the extents of 'number of products in a catalog' and/or 'number of catalogs a product is in' are, we may also want to consider other storage models for the collection, including some kind of paginating structure similar to the sensor record storage in track and trace.

I've considered an implementation like this, the intention around modeling catalog in its current state was to 'mimic' the state of software catalogs today. I am not strongly opinionated in either approach. Though it would probably warrant a design discussion on this implementation to ensure we're all on the same page. 

Perhaps GS1 would be able to comment on whether the cardinality of products to catalogs is materially higher than catalogs to products. We can do some internal research ourselves

Reviewer 2: I think I need to understand more about the overall approach.  Today,  there typically is 1 supplier catalog (as it is today), all products listed, with certain products shared with specific trading partners? So - generally more products than catalogs. Numbers of products are exponentially higher than catalogs. HOWEVER - are you guys thinking there is a 1:1 catalog for every trading partner and the products shared between those parties constitutes the catalog (NOT how it works today)

Author: Good to note that suppliers have one "master" catalog and share specific products with trading partners. Perhaps an introduction of a catalog_type enum would make sense. A "MASTER_CATALOG" could be the all-encompassing catalog of products, while a "TRADING_PARTNER_CATALOG" could be a subset of those products with specific prices for the trading partner.

Property Schema Transaction Family and are restricted to the fields and rules of
the GS1 Catalog schema (which will be defined at a later time). Transactions
which are responsible for setting product state values must ensure that those
properties conform with the requirements of the GS1 Catalog Property Schema (to
be defined at a later time).

``` 
message Catalog { 
    string catalog_id = 1; 
    string owner = 2; 
    string expiry_date = 3; 
    name = 4; 
    ProductList products = 5; 
    repeated PropertyValue properties = 6; 
} 
```

# Rationale and alternatives
[alternatives]: #alternatives

The design of catalog in this manner leaves it extensible for domain specific
use. The concept of price has been discussed and it's been decided to include a
notion of price within products added to a catalog. A simple implmentation of
price is fine and currently serves as a placeholder, while a more robust
standard of modeling price data can be implemented from a future pricing-related
RFC. A GS1 pricing standard that can be leveraged in the writing of such an RFC
would be [GS1 Price
Sync](https://www.gs1.org/docs/gdsn/3.1/BMS_Price_Sync_r3p1p3_i1p3p5_23May2017.pdf).

> 

Reviewer 2: Understanding that pricing is variable by trading partner relationship for the same product, how would this be implemented? A separate record for the same item  - tied to the partner the supplier may be doing business with? Goes back to the question above - is the intent to have a 1:1 catalog for every trading partner relationship or model more what is today which is a 1:many?

Author: First question:
From the Rational and Alternatives section: 
> The concept of price has been discussed and it's been decided to include
the notion of price within products added to a catalog. A simple implementation of
price is fine and currently serves as a placeholder, while a more robust
standard of modeling price data can be implemented from a future pricing-related
RFC.

Second question: Yes, we could require something like a trading_partner_id for catalogs of type "TRADING_PARTNER_CATALOG". 

Third question: Addressed above

Reviewer: I don't know if it makes sense to sub-type these things. I think we should keep it loose and allow the creating organizations to define what it is. This may be very supplier (catalog-creator) specific. For example, an organization might have a single customer in a given segment and may publish a unique catalog of products directed at that customer. Then, in the future, they might acquire a different customer in that same segment and decide that catalog is suitable for providing to them as well. Ultimately, the segmentation is based on choices the creators make and what parties they choose to share the data with.

Reviewer 3: The flexibility feels really important.

The supplier may have contractual obligations with the trading partner to keep some/all catalog items in a private context (however that might work) - probably more B2B.

Perhaps the supplier having a "flag" that allows catalogs/products/items to be "listed" as B2B, B2C, etc., for traditional commerce allows some "logic" flow to be identified for other commerce types, things like Business2Government.

Author: > The flexibility feels really important.
> 
> The supplier may have contractual obligations with the trading partner to keep some/all catalog items in a private context (however that might work) - probably more B2B.
> 
> Perhaps the supplier having a "flag" that allows catalogs/products/items to be "listed" as B2B, B2C, etc., for traditional commerce allows some "logic" flow to be identified for other commerce types, things like Business2Government.

I see where you're coming from. I think in the context of "may" have contractual obligations, a "flag" field for this would be better included as an additional property_value enforced by a grid schema that could be a fork of the catalog_product schema.

# Prior art
[prior-art]: #prior-art

In the distributed ledger space there are not any example implementations of a
product catalog. There are specs and standards around catalogs that can be
leveraged in the design and implementation of a Grid Catalog.

See Ariba's® [catalog interchange format
(CIF)](https://www.essent.com/What-is-a-CIF-Catalog.html) CIF that has grown
into a de facto standard and is widely used by hundreds of thousands of
companies throughout the world.

[GS1's Catalogue Item
Synchronisation](https://www.gs1.org/docs/gdsn/3.1/BMS_GDSN_Catalogue_Item_Sync_r3p1p0_i1_p0_p6_25Aug2015.pdf)
is the process of continuous harmonisation of item information between trading
partners which ensures that the master data is the same in all trading partners
systems.

> 

Reviewer 2: This is one spec, but is part of a much larger data synchronization network.  Can we get more information about the hundreds of thousands of companies using CIF and in what manner for what level of product? I went to the link but did not see users, or industry groups

Author: AFAIK it's something baked into SAP -- @davececchi may know more about it than I do.

I'll update the link, this has more details about CIF: https://public.cummins.com/sites/CSP/SiteCollectionDocuments/Help/AribaNetworkHelp/CIF_Catalog_Guide.pdf

- Feature Name: Catalog
- Start Date: 2019-07-12
- RFC PR: 
- Hyperledger Grid Issue: [GRID-95](https://jira.hyperledger.org/browse/GRID-95)

# Summary
[summary]: #summary

This RFC proposes a generic implementation for a Hyperledger Grid *Catalog*. The
implementation of *Catalog* will materialize as an assortment of grid products
that can be shared with one or more organizations, as well as implementaion of
catalog related functions. The Catalog on grid will contain 5 fields. A

> **catalog_id**, a **catalog_owner**, an **expiry_date**, a **Name** for the 

Reviewer 4: ```suggestion
**catalog_id**, a **catalog_owner**, an **expiry_date**, a **name** for the 
```

catalog, and a repeated field of PropertyValues for custom fields.

In addition any grid product added to a catalog will need to contain a
***status*** field and can include an optional ***prices*** field, which will be 
enforced by the catalog_product grid schema.

# Motivation
[motivation]: #motivation

The Grid Catalog implementation is designed for sharing an assortment of
products between organizations. Each product in a catalog will contain a number
of master data elements. Using the Grid Catalog, organizations can share
collections of related/unrelated products with other participants in their
network. 

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## Entities

> A **catalog** is a high level construct that products adhereing to the 

Reviewer 4: ```suggestion
A **catalog** is a high-level construct that products adhering to the 
```

"catalog_product" schema can reference as an additional field.  A catalog can be 
referenced, or used to do business in a supply chain. A catalog is referenced by 
a "catalog_product" via its **catalog_id**. A catalog also has an **owner** (the 
organization that creates the catalog).  It also contains an **expiry_date** as 

> a mechanism for clients to exipire the catalog. Also a **Name** for the catalog. 

Reviewer 4: ```suggestion
a mechanism for clients to expire the catalog. Also a **Name** for the catalog. 
```

organization in the catalog's owner field.  An organization must have the GS1
company prefix in the "gs1_company_prefixes" metadata field for its Pike
organization.  (Organization-level metadata will need to be implemented in
Pike.) When a catalog is created, its owning organization is stored with the
catalog in an "owner" field.

Deletion of a catalog is restricted to agents acting on behalf of the
organization in the catalog's owner field.  A setting will turn off deletion
entirely, with the intent that it could be enabled/disabled by a Grid
administrator.  Disabling delete is useful because external systems may have
references to these products and deleting them could leave dangling references.

*Disclaimer: CatalogDelete is a potentially hazardous operation and needs to be 
done with care.*

Replication of a catalog is restricted to agents acting on behalf of an
organization. The purpose of replication is to enable agents to duplicate a 
catalog, or a subset of "catalog_products" in state with the intention of 
modifying the catalog_products to fit a different scheme. It reduces the 
overhead needed to recreate a catalog. Generally organizations have a single 

> "master" catalog that they use to build "subcatalogs" customized for their 

Reviewer 4: ```suggestion
"master" catalog that they use to build "sub-catalogs" customized for their 
```

To perform any of the **catalog actions** agents will need the following 
permissions:
```
CatalogCreate: Agent from org needs the "can_create_catalog" permission
CatalogDelete: Agent from org needs the "can_delete_catalog" permission
CatalogReplicate: Agent from org needs the "can_duplicate_catalog" permission
```

In the case of **catalog operations** agents from the org they represent will
need the operational permission to perform the desired operation on a Grid
Catalog.
```
AddProductsToCatalog: Agent from org needs the "can_add_products_to_catalog" permission
RemoveProductsFromCatalog: Agent from org needs the "can_remove_products_from_catalog" permission
ActivateProduct: Agent from org needs the "can_activate_product_in_catalog" permission
DeactivateProduct: Agent from org needs the "can_deactivate_product_in_catalog" permission
DiscontinueProduct: Agent from org needs the "can_discontinue_product_in_catalog" permission
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The primary object stored in state is **Catalog**, which consists of a
**catalog_id** ([UUID](https://crates.io/crates/uuid)), an **owner** (org_id 

> compatible w/Pike), an **expiry_date** (unix timestamp), a **Name**, and a 

Reviewer 4: ```suggestion
compatible w/Pike), an **expiry_date** (unix timestamp), a **name**, and a 
```

repeated field of **PropertyValues**.  The properties available are defined by 
the Grid Property Schema Transaction Family and are restricted to the fields and 
rules of the GS1 Catalog schema (which will be defined at a later time).  
Transactions which are responsible for setting product state values must ensure 
that those properties conform with the requirements of the GS1 Catalog Property 
Schema (to be defined at a later time).

``` 
message Catalog { 
    string catalog_id = 1; 
    string owner = 2; 
    string expiry_date = 3; 
    name = 4; 
    repeated PropertyValue properties = 6; 
} 
```

### Referencing a Catalog

> Products are uniquely referenced by their product_id and product_namespace.  For

Reviewer 4: I think this is a copy/paste from the Product RFC

GS1, Products are referenced by the GTIN identifier. For example:

``` 
get_catalog(catalog_id) // UUID 
set_catalog(catalog_id, catalog) // UUID, catalog 
```

### Catalog Addressing in the Merkle-Radix State System

In order to uniquely locate Catalogs in the Merkle-Radix state system, an
address must be constructed which identifies the storage location of the Catalog
representation.

All Grid addresses are prefixed by the 6-hex-character namespace prefix
"621dee",  Catalogs are further prefixed under the Grid namespace with reserved
enumerations of "03" ("00", "01", and "02" being reserved for other purposes)
indicating "Catalogs" and an additional "01" indicating "GS1 Catalog".

Therefore, all addresses starting with:

``` 
"621dee" + "03" + "01" 
```

> are Grid GS1 Catalogs identified by a UUID which can be reference by a 

Reviewer 4: ```suggestion
are Grid GS1 Catalogs identified by a UUID which can be referenced by a 
```

"catalog_product" to group products to a specific catalog. 

UUID formats consist of 36-digit "alphanumeric strings" which include a fixed 
amount of internal "0" padding.  After the 10-hex-characters that are consumed 
by the grid namespace prefix, the catalog, and GS1 prefixes, there are 60 hex 

> characters remaining in the address.  The 36 digits of the GTIN can be left 

Reviewer 4: GTIN's aren't 36 characters, right?

Author: correct

padded with 24-hex-character zeroes and right padded with 2-hex-character zeroes 
to accommodate potential future storage associated with the GS1 Catalog
representation, for example:

``` 
"621dee" + "03" + "01" + 0000000000000000000000 +
36-character "numeric string" catalog_id + "00" // catalog_id == UUID 
```

Using the v5 constructer in the UUID crate we can tie the catalog_id to 

> catalog_name will prevent duplication of catalogs when invoking the 

Reviewer 4: ```suggestion
catalog_name to prevent duplication of catalogs when invoking the 
```

catalog_replicate action.  
```
use uuid::Uuid;

fn main() {
    let my_uuid = Uuid::new_v5(&Uuid::new_v4(), b"catalog_name");
    println!("{}", my_uuid);
}

>> 9db5fcec-dcab-5b04-b31b-765ccbf2bc3a
```

The full catalog address would look like:
``` 

> "621dee030100000000000000000000009db5fcec-dcab-5b04-b31b-765ccbf2bc3a00" 

Reviewer 4: Hyphens aren't a valid hexadecimal character so would this actually look like `621dee030100000000000000000000009db5fcecdcab5b04b31b765ccbf2bc3a00`?

```
Schema(
    name="Catalog Product",
    description="A grid product that can be grouped with other products via grid catalog",
    owner = "Target"
    properties=[
        PropertyDefinition(
            name="status",
            data_type=PropertyDefinition.DataType.ENUM,
            description="The current state of the catalog product",
            enum_options=["ACTIVE", "INACTIVE"])
            required=True
        ),
        PropertyDefinition(
            name="price",
            data_type=PropertyDefinition.DataType.STRING,
            description="The price of the catalog product"
            required=False
        )])
```

In order to store a catalog product in global state, we will extend from the definition of a grid product with several fixed fields, namely identifiers, as well as the set of dynamic property values that will conform to the catalog product shema definition.

```
message Product { 
    enum ProductNamespace { 
        UNSET_NAMESPACE = 0; 
        GS1 = 1; 
    }
    
    ProductNamespace product_namespace = 1; 
    string product_id = 2; 
    string owner = 3;
    repeated PropertyValue properties = 4; 
}
```

> The catalog product smart contract would then be responsible for validating the properties field against the Lightbulb schema at run-time.

Reviewer 4: "Lightbulb" schema?

> An Catalog Product entry, with all optional properties, would then look like the following:

Reviewer 4: ```suggestion
A Catalog Product entry, with all optional properties, would then look like the following:
```

Schema (to be defined at a later time).

``` 
message Catalog { 
    string catalog_id = 1; 
    string owner = 2; 
    string expiry_date = 3; 
    string name = 4; 
    repeated PropertyValue properties = 5; 
} 
```

### Referencing a Catalog

Catalogs are uniquely referenced by their catalog_id. For example:

``` 
get_catalog(catalog_id) // hash(catalog_name) 
set_catalog(catalog_id, catalog) // hash(catalog_name), catalog 
```

### Catalog Addressing in the Merkle-Radix State System

In order to uniquely locate Catalogs in the Merkle-Radix state system, an
address must be constructed which identifies the storage location of the Catalog
representation.

All Grid addresses are prefixed by the 6-hex-character namespace prefix
"621dee",  Catalogs are further prefixed under the Grid namespace with reserved
enumerations of "03" ("00", "01", and "02" being reserved for other purposes)
indicating "Grid Catalogs" and an additional org_id. Currently in Grid Product, org_ids are a [6-10 digit GS1 company prefix](https://www.gs1-us.info/gs1-company-prefix/). This composite address will enable inexpensive catalog operations.

Therefore, all addresses starting with:

``` 

> "621dee" + "03" + "123456" // org_id (in this case a gs1_company_prefix)

Author: left pad to 10 digits using "a"

the catalog_product grid schema.

```
Schema(
    name="Catalog Product",
    description="A grid product that can be grouped with other products via grid catalog",
    owner = "Target"
    properties=[
        PropertyDefinition(
            name="catalog_id",
            data_type=PropertyDefinition.DataType.STRING,
            description="The the id of the catalog this 'catalog product' belongs to",
            required=True
        ),
        PropertyDefinition(
            name="status",
            data_type=PropertyDefinition.DataType.ENUM,
            description="The current state of the catalog product",
            enum_options=["ACTIVE", "INACTIVE"],
            required=True
        ),
        PropertyDefinition(
            name="price",
            data_type=PropertyDefinition.DataType.STRING,
            description="The price of the catalog product"
            required=False
        )])
```

In order to store a catalog product in global state, we will extend from the 
definition of a grid product with several fixed fields, namely identifiers, as 
well as the set of dynamic property values that will conform to the catalog 
product schema definition.

```

> message Product { 

Reviewer 5: This doesn't match the example below

Author: ✅

Reviewer 5: I think I understand what I was missing here: you referenced Product from the product RFC, and the give an example of `CatalogProduct`, but haven't defined the proto for `CatalogProduct`.

Took me a bit to figure that out because one of the fields in the example has some whitespace/formatting issues that made it look like it was part of a trailing comment.

    enum ProductNamespace { 
        UNSET_NAMESPACE = 0; 
        GS1 = 1; 
    }
    
    ProductNamespace product_namespace = 1; 
    string product_id = 2; 
    string owner = 3;
    repeated PropertyValue properties = 4; 
}
```

The catalog smart contract would then be responsible for validating the 

> properties field against the CatalogProduct schema at run-time.

Author: Add note about owner inheritance from product -> catalogproduct

A catalog product entry, with all optional properties, would then look like the following:
```
CatalogProduct(
    product_id="gtin",
    product_namespace="GS1"
    owner='Target',
    properties=[
        PropertyDefinition(
            name="catalog_id",
            data_type=PropertyDefinition.DataType.STRING,
            string_value="7d1734adfee4927"
        ),
        PropertyValue(
            name="status",
            data_type=PropertyDefinition.DataType.ENUM,
            enum_value=1, # ACTIVE (default)
        ),
        PropertyValue(
            name="price"
            data_type=PropertyDefinition.DataType.STRING,
            string_value="1-10:$50, 10-50:$40, 50+:$30"  # price windows
        ),
    ])
```

## Transaction Payload and Execution (catalog operations)

CatalogPayload Transaction
CatalogPayload contains an action enum and the associated action payload. This allows for the action payload to be dispatched to the appropriate logic.

Only the defined actions are available and only one action payload should be defined in the CatalogPayload.

> message CatalogPayload { 

Reviewer 5: missing codeblock markup around this.

Author: ✅

    enum Actions { 
        UNSET_ACTION = 0; 
        CATALOG_CREATE = 1; 
        CATALOG_DELETE = 2; 
        CATALOG_REPLICATE = 3; 
    }

    Action action = 1;

    // Approximately when transaction was submitted, as a Unix UTC timestamp
    uint64 timestamp = 2;

    CatalogCreateAction catalog_create = 3; 
    CatalogDeleteAction catalog_delete = 4; 
    CatalogReplicateAction catalog_replicate = 5; 
} 

### CatalogCreateAction
CatalogCreateAction adds a new catalog to state. The transaction should be submitted by an agent, which is identified by its signing key, acting on behalf of the organization that corresponds to the owner in the create transaction. (Organizations and agents are defined by the Pike smart contract.)

> message CatalogCreateAction { 

Reviewer 5: missing codeblock markup

Author: ✅

    // a composite key for determining the state address
    string owner = 1;
    string catalog_id = 2;
    string name = 3;
    string expiry_date = 4;
    repeated PropertyValues properties = 4; 
} 

Validation requirements:

- If a catalog with catalog_id already exists the transaction is invalid.
- The signer of the transaction must be an agent in the Pike state and must belong to an organization in Pike state, otherwise the transaction is invalid.
- The agent must have the permission can_create_catalog for the organization, otherwise the transaction is invalid.
- If the product_namespace is GS1, the organization must contain a GS1 Company Prefix in its metadata (gs1_company_prefixes), and the prefix must match the company prefix in the product_id, which is a GTIN if GS1, otherwise the transaction is invalid.
- The properties must be valid for the product_namespace. For example, if the product is GS1 product, its properties must only contain properties that are included in the GS1 Schema. If it includes a property not in the GS1 Schema the transaction is invalid.

AddProductsToCatalog
RemoveProductsFromCatalog
ActivateProduct
DeactivateProduct
DiscontinueProduct

> # Rationale and alternatives

Author: Add a note about why we are using a reference-based implementation as opposed to a list concept

be left padded with the hexidecimal representation of the number 10 ("a") until it is 
10 digits in length. This composite address will enable inexpensive catalog operations.

Therefore, all addresses starting with:

``` 
"621dee" + "03" + "aaaa123456" // org_id (in this case a gs1_company_prefix)
```

are Grid GS1 Catalogs identified by the hash of the catalog name which can be
referenced by a "catalog_product" to group products to a specific catalog. 

The catalog_id format consists of 15-digit "alphanumeric string" which include 
a fixed amount of internal "0" padding.  After the 18-hex-characters that are 
consumed by the grid namespace prefix, the catalog, and gs1_company_prefix there 
are 52 hex characters remaining in the address.  The 15 digits of the catalog_id can 
be left padded with 60-hex-character zeroes and right padded with 
2-hex-character zeroes to accommodate potential future storage associated with 
the GS1 Catalog representation, for example:

``` 
"621dee" + "03" + "123456" + "000000000000000000000000000000000000000" +
15-character "numeric string" catalog_id + "00" // catalog_id == hash(catalog_name) 
```

> Using SHA-3 from the rust-crypto crate we can tie the catalog_id to 

Author: ✅

- Feature Name: Catalog
- Start Date: 2019-07-12
- RFC PR: 
- Hyperledger Grid Issue: [GRID-95](https://jira.hyperledger.org/browse/GRID-95)

# Summary
[summary]: #summary

This RFC proposes a generic implementation for a Hyperledger Grid *Catalog*. The
implementation of *Catalog* will materialize as an assortment of grid products

> that can be shared with one or more organizations, as well as implementation of

Reviewer 6: Based on our discussion we agreed that the product properties should only be core product attributes and that all context specific attributes should be defined via a context specific schema. the context could be a specific retailer, a region or maybe some seasonal events.

Deletion of a catalog is restricted to agents acting on behalf of the
organization in the catalog's owner field.  A setting will turn off deletion
entirely, with the intent that it could be enabled/disabled by a Grid
administrator.  Disabling delete is useful because external systems may have
references to these products and deleting them could leave dangling references.

*Disclaimer: CatalogDelete is a potentially hazardous operation and needs to be 
done with care.*

Replication of a catalog is restricted to agents acting on behalf of an
organization. The purpose of replication is to enable agents to duplicate a 
catalog, or a subset of "catalog_products" in state with the intention of 
modifying the catalog_products to fit a different scheme. It reduces the 
overhead needed to recreate a catalog. Generally organizations have a single 
"master" catalog that they use to build "sub-catalogs" customized for their 
business partners. 

To perform any of the **catalog actions** agents will need the following 
permissions:
```
CatalogCreate: Agent from org needs the "can_create_catalog" permission
CatalogDelete: Agent from org needs the "can_delete_catalog" permission

> CatalogReplicate: Agent from org needs the "can_duplicate_catalog" permission

Reviewer 6: agreed

namespace prefix, the product namespace prefix, GS1 namespce prefix, and catalog 
product namespace prefix, there are 58 hex characters remaining in the address.  
The 12 to 14 digits of the GTIN can be left padded with 40 to 42-hex-character zeroes and 
right padded with 2-hex-character zeroes to accommodate potential future storage 
associated with the GS1 Product representation, for example:

``` 
“621dee” + “02” + “01” + "846e4a25c881f56" + “00000000000000000000000000000” 
+ 14-character “numeric string” product_id + “00” // product_id == GTIN 
```

A full Catalog Product address using the example GTIN from https://www.gtin.info/ would therefore be:

``` 
“621dee0201846e4a25c881f56000000000000000000000000000000001234560001200” 
```

### Catalog Product Schema Definition

This schema defines the additional attributes required of a grid product, such that 
it may be considered a "catalog product". Catalog products should be grouped with 
other catalog products. Once grouped, the grid catalog can be shared with other 
participants in a grid network.

Any grid product added to a catalog will need to contain a _required_ ***status*** 

> field and can include an _optional_ ***prices*** field, which will be enforced by 

Author: The associated _schema_ is applied at a _catalog_product_ level

- Feature Name: Catalog
- Start Date: 2019-07-12
- RFC PR: 
- Hyperledger Grid Issue: [GRID-95](https://jira.hyperledger.org/browse/GRID-95)

# Summary
[summary]: #summary

This RFC proposes a generic implementation for a Hyperledger Grid *Catalog*. The
implementation of *Catalog* will materialize as a reference based ID shared between 
instances of catalog_products. That catalog_id can be used to share a grouping of 
products with that can be shared with one or more organizations. As well as assist 
in programmatic operations performed on a Catalog. The Grid Catalog will contain 4 
fields. A **catalog_id**, a **catalog_owner**, a **Name** for 
the catalog, and a repeated field of PropertyValues for custom fields.

In addition an example catalog product schema is included as a way to demonstrate 
enforcement of additional product properties. Any grid product added to a catalog 

> will adhere to the "catalog_product" schema (or another custom-defined schema). The 

Author: Add note about attributes that go into "catalog_product" will be trading partner relationship-specific attributes like price, return policy, etc. Where the higher leve Grid Product will contain the base GS1 attributes (weights, dims, etc.)

Author: ✅

- Feature Name: Catalog
- Start Date: 2019-07-12
- RFC PR: 
- Hyperledger Grid Issue: [GRID-95](https://jira.hyperledger.org/browse/GRID-95)

# Summary
[summary]: #summary

This RFC proposes a generic implementation for a Hyperledger Grid *Catalog*. The

> implementation of *Catalog* will materialize as a reference based ID shared bewtween 

Reviewer 3: `s/bewtween/between`

Author: ✅

> instances of of grid products. That catalog ID can be used to share a grouping of 

Reviewer 3: remove one of the `of`s

Author: ✅

products with that can be shared with one or more organizations. As well as assist 
in programmatic operations performed on a Catalog. The Grid Catalog will contain 5 
fields. A **catalog_id**, a **catalog_owner**, an **expiry_date**, a **Name** for 
the catalog, and a repeated field of PropertyValues for custom fields.

In addition an example "catalog_product" is included as a way to demonstrate 
schema enforcement of additional product properties. Any grid product added to a 
catalog will adhere to the "catalog_product" schema (or another custom-defined 
schema) which requires a ***catalog_id*** field, a***status*** field and an 
optional ***prices*** field, which will be enforced by the catalog_product grid 
schema.

> A **catalog_id** field ***must*** be included in any catalog related schema. All other 

Author: move catalog_id as a top level "catalog_product" attribute

Author: ✅

> - Feature Name: Catalog

Author: ✅

- Start Date: 2019-07-12
- RFC PR: 
- Hyperledger Grid Issue: [GRID-95](https://jira.hyperledger.org/browse/GRID-95)

# Summary
[summary]: #summary

> This RFC proposes a generic implementation for a Hyperledger Grid *Catalog*. The

Author: ✅

> implementation of *Catalog* will materialize as a reference based ID shared between 

Author: ✅

instances of catalog_products. That catalog_id can be used to share a grouping of 

> products with that can be shared with one or more organizations. As well as assist 

Author: ✅

> in programmatic operations performed on a Catalog. The Grid Catalog will contain 4 

Author: Removed ✅

fields. A **catalog_id**, a **catalog_owner**, a **Name** for 
the catalog, and a repeated field of PropertyValues for custom fields.

In addition a base catalog product schema is included as a way to demonstrate 
enforcement of additional product properties. Any grid product added to a catalog 
will adhere to the "catalog_product" schema (or another custom-defined schema). The 
example schema in particular will enforce a required ***status*** field, an optional 
***return_policy***, and an optional ***price*** field.

# Motivation
[motivation]: #motivation

The Grid Catalog implementation is designed for sharing an assortment of
products between organizations. Each product in a catalog will contain a number
of master data elements. Using the Grid Catalog, organizations can share
collections of related/unrelated products with other participants in their
network. 

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## Entities

> A **catalog** is a high-level construct that catalog_products can reference. The "catalog_product" 

Author: Added a definition of catalog_product ✅

> schema can be extended to include any number of attributes that are required by an organization. A catalog can be referenced, or used to share item level information in a supply chain. 

Author: ✅

A catalog is referenced by a "catalog_product" via its **catalog_id**. A catalog also 
has an **owner** (the organization that creates the catalog).  Also a **Name** for 
the catalog. 

> A catalog, like a product, can also have one or more **properties**.  Properties

Author: added a note ✅

are described in the Grid Primitives RFC.  A *property namespace* contains
multiple *property schemas*.  A property schema associates a name (such as
"length") with a data type (such as integer).  

## Transactions

Catalogs are managed by submitting transactions to Hyperledger Grid, which will
process them with the Grid Catalog smart contract. The following transactions
are supported:

**catalog actions:**
* CatalogCreate - create a catalog and store it in state
* CatalogUpdate - update a catalog in state
* *CatalogDelete - deletes a catalog from state and associated catalog_products

**catalog_product actions:**
* CatalogProductCreate - create a catalog_product and stores it in state
* CatalogProductUpdate - updates a catalog_product in state
* CatalogProductDelete - delete a catalog_product from state

**catalog_product operation:**

> * *SetCatalogProductStatus - changes the "status" of a catalog_product to an active, inactive, or discontinued state

Author: ✅

references to these products and deleting them could leave dangling references.

*Disclaimer: CatalogDelete & CatalogProductDelete are potentially hazardous operation and need to be done with care.*

To perform any of the **catalog actions** agents will need the following 
permissions:
```
CatalogCreate: Agent from org needs the "can_create_catalog" permission
CatalogUpdate: Agent from org needs the "can_update_catalog" permission
CatalogDelete: Agent from org needs the "can_delete_catalog" permission
```

To perform any of the **catalog_product actions** agents inherit the same 
permissions needed for a Grid Product:
```
CatalogProductCreate: Agent from org needs the "can_create_product" permission
CatalogProductUpdate: Agent from org needs the "can_update_product" permission
CatalogProductDelete: Agent from org needs the "can_delete_product" permission
```

In the case of **catalog operations** agents from the org they represent will
need the operational permission to perform the desired operation on a Grid
Catalog.
```
ActivateCatalogProduct: Agent from org needs the "can_activate_product_in_catalog" permission

> DeactivateCatalogProduct: Agent from org needs the "can_deactivate_product_in_catalog" permission

Author: Agreed

> DeactivateCatalogProduct: Agent from org needs the "can_deactivate_product_in_catalog" permission

Author: Consolidated to 1 permission

> DeactivateCatalogProduct: Agent from org needs the "can_deactivate_product_in_catalog" permission

Reviewer 7: This comment was not a change request

> DeactivateCatalogProduct: Agent from org needs the "can_deactivate_product_in_catalog" permission

Author: Gotcha. This could be better consolidated as one permission. If a higher level of granularity around permissions for adjusting a product's status is required, we can update it in a future iteration. For now, I think it's fine

DiscontinueProduct: Agent from org needs the "can_discontinue_product_in_catalog" permission
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## State

### Catalog Representation

The primary object stored in state is **Catalog**, which consists of a
**catalog_id** (hash of the catalog_name), an **owner** (org_id 
compatible w/Pike), a **Name**, and a 
repeated field of **PropertyValues**.  The properties available are defined by 
the Grid Property Schema Transaction Family and are restricted to the fields 

> and rules of the GS1 Catalog schema (which will be defined at a later time).  

Reviewer: I feel like we should take a crack at what this looks like.

> Transactions which are responsible for setting product state values must ensure 

Author: ✅

```

### Catalog Addressing in the Merkle-Radix State System

In order to uniquely locate Catalogs in the Merkle-Radix state system, an
address must be constructed which identifies the storage location of the Catalog
representation.

All Grid addresses are prefixed by the 6-hex-character namespace prefix
"621dee",  Catalogs are further prefixed under the Grid namespace with reserved
enumerations of "03" ("00", "01", and "02" being reserved for other purposes)
indicating "Grid Catalogs" and an additional org_id. Currently in Grid Product, 
org_ids are a [6-10 digit GS1 company prefix](https://www.gs1-us.info/gs1-company-prefix/). 
To keep the prefix length consistent, any GS1 company prefix less than 10 digits will 
be left padded with the hexadecimal representation of the number 10 ("a") until it is 
10 digits in length. This composite address will enable inexpensive catalog operations.

Therefore, all addresses starting with:

``` 
"621dee" + "03" + "aaaa123456" // org_id (in this case a gs1_company_prefix)
```

are Grid GS1 Catalogs identified by the hash of the catalog name which can be
referenced by a "catalog_product" to group products to a specific catalog. 

> The catalog_id format consists of 15-digit "alphanumeric string" which include 

Reviewer: s/consists of 15-digit/consists of a 15-digit/

> The catalog_id format consists of 15-digit "alphanumeric string" which include 

Reviewer: s/which include/which includes/

a fixed amount of internal "0" padding.  After the 18-hex-characters that are 
consumed by the grid namespace prefix, the catalog, and gs1_company_prefix there 
are 52 hex characters remaining in the address.  The 15 digits of the catalog_id can 
be left padded with 60-hex-character zeroes and right padded with 
2-hex-character zeroes to accommodate potential future storage associated with 
the GS1 Catalog representation, for example:

``` 

> "621dee" + "03" + "123456" + "000000000000000000000000000000000000000" +

Author: ✅  Also fixed some wording this that paragraph

15-character "numeric string" catalog_id + "00" // catalog_id == hash(catalog_name) 
```

Using SHA-2 from the rust-crypto crate we can tie the catalog_id to 
catalog_name to prevent duplication of catalogs when invoking the 
catalog_create action.  
```
extern crate crypto;
use self::crypto::digest::Digest;

> use self::crypto::sha3::Sha3;

Reviewer 8: Current code (pike, product) uses crypto::sha2::Sha512. We should probably stick to the same one.

Author: ✅

fn main(){
    // create a SHA2-256 hasher
    let mut hasher = Sha256::new();

    // write input message
    hasher.input_str("catalog_name");

    // read hash digest
    let hex = hasher.result_str();
    let hex_15 = &hex[0..15];

    println!("SHA2-256 Hash is {:?}", hex);
    println!("SHA2-256 Hash (15) {:?}", hex_15);
}

>> SHA3 Hash is: e70fe351d1d096c263d041624c4524da22693cabeef978c6e1d9b1b0dc266a65
>> SHA3 Hash (15): e70fe351d1d096c
```

The full catalog address would look like:
``` 

> "621dee03123456000000000000000000000000000000000000000e70fe351d1d096c00" 

Author: ✅

```

### Referencing a Catalog Product

Catalogs products are uniquely referenced by a composite key composed of their catalog_id 
and product_id. For example:

``` 
get_catalog_product(catalog_id, product_id) // hash(catalog_name), product_id
set_catalog_product(catalog_id, product_id, catalog_product) // hash(catalog_name), product_id, 
catalog_product
```

### Catalog Product Addressing in the Merkle-Radix State System

The Grid GS1 Catalog Product state address will be prefixed with the Grid 
namespace of "621dee", in addition to "02" (product namespace), "01" for (gs1 
product namespace), and the catalog_id.

Therefore, all addresses starting with:

``` 
"621dee" + "02" + "01 + "e70fe351d1d096c" // catalog_id
```

are Grid GS1 Catalog Products and are expected to contain the standard 
attributes of a Grid GS1 Product as well as the extra PropertyValues that 

> are defined in the  catalog product schema. These PropertyValues are in 

Reviewer: extra space

GTIN-13, or GTIN-14).  After the 12-hex-characters that are consumed by the grid
namespace prefix, the product namespace prefix, GS1 namespace prefix, and catalog 
product namespace prefix, there are 58 hex characters remaining in the address.  
The 12 to 14 digits of the GTIN can be left padded with 40 to 42-hex-character 
zeroes and right padded with 2-hex-character zeroes to accommodate potential 
future storage associated with the GS1 Product representation, for example:

``` 
"621dee" + "02" + "01" + "e70fe351d1d096c" + "00000000000000000000000000000" 
+ 14-character "numeric string" product_id + "00" // product_id == GTIN 
```

A full Catalog Product address using the example GTIN from https://www.gtin.info/ would therefore be:

``` 
"621dee0201e70fe351d1d096c000000000000000000000000000000001234560001200" 
```

### Catalog Product Schema Definition

This schema defines the additional attributes required of a grid product, such that 
it may be considered a "catalog product". Catalog products should be grouped with 
other catalog products. Once grouped, the grid catalog can be shared with other 
participants in a grid network.

> 

Reviewer: Extra linefeed

Any grid product added to a catalog (thus becoming a catalog_product) will need to 

> contain a reference fields to the original product (product_id, product_namespace, 

Reviewer: s/reference fields/reference/

and owner) and a catalog_id. A catalog product schema may also be defined to 
required additional fields for that catalog_product. This example schema enforces 
a _required_ ***status*** field and an _optional_ ***prices*** field.

Example schema:
```
Schema(
    name="Catalog Product",
    description="A grid product that can be grouped with other products via grid catalog",
    owner = "Target"
    properties=[
        PropertyDefinition(
            name="status",
            data_type=PropertyDefinition.DataType.ENUM,
            description="The current state of the catalog product",
            enum_options=["ACTIVE", "INACTIVE"],
            required=True
        ),
        PropertyDefinition(
            name="price",
            data_type=PropertyDefinition.DataType.STRING,
            description="The price of the catalog product"
            required=False
        ),
        PropertyDefinition(
            name="return_policy",
            data_type=PropertyDefinition.DataType.STRING,
            description="The return policy of the of the catalog product"
            required=False
        )])
```

> In order to store a catalog product in global state, we will extend from the 

Reviewer: s/extend from/extend/

**Grid Product from the [Product RFC](https://github.com/hyperledger/grid-rfcs/blob/master/text/0005-product.md):**

```
message Product { 
    enum ProductNamespace { 
        UNSET_NAMESPACE = 0; 
        GS1 = 1; 
    }
    
    ProductNamespace product_namespace = 1; 
    string product_id = 2; 
    string owner = 3;
    repeated PropertyValue properties = 4; 
}
```

The catalog_product smart contract would then be responsible for validating the 
properties field against the CatalogProduct schema at run-time.

A catalog_product entry would reference Grid Product. In addition to 
the properties defined in the CatalogProduct schema. The data model 
would look like this:
```
CatalogProduct(
    product_id="gtin", // Reference from Grid Product
    product_namespace="GS1" // Reference from Grid Product
    owner='Target', // Reference from Grid Product
		catalog_id='e70fe351d1d096c'
    properties=[
        PropertyValue(
            name="status",
            data_type=PropertyDefinition.DataType.ENUM,
            enum_value=1, # ACTIVE (default)
        ),
        PropertyValue(
            name="price"
            data_type=PropertyDefinition.DataType.STRING,

>             string_value="1-10:$50, 10-50:$40, 50+:$30"  # price windows

Reviewer: Maybe make a comment here that real structured data for breakpoint discounts would likely be modeled in a schema, rather than this string example.

        ),
        PropertyValue(
            name="return policy"
            data_type=PropertyDefinition.DataType.STRING,
            string_value="30 days from sale"
        ),
    ])
```

## Transaction Payload and Execution 

The following payloads contain an action enum and the associated action payload. 
This allows for the action payload to be dispatched to the appropriate logic.
Only the defined actions (enum) are available, and only one Action should be 
performed in a payload.

### CatalogPayload Transaction

```

> message CatalogPayload { 

Reviewer 7: Are each of these payloads going to be handled in different smart contracts? If they are all apart of one it would be better to combine them into on payload. To provide more logical room for each section to have more actions you can separate them by have the actions for catalog start at 1, catalog product at 100  etc.

Author: Consolidated into one payload ✅

        UNSET_ACTION = 0; 
        CATALOG_PRODUCT_CREATE = 1; 
        CATALOG_PRODUCT_UPDATE = 2; 
        CATALOG_PRODUCT_DELETE = 3; 
    }

    Action action = 1;

    // Approximately when transaction was submitted, as a Unix UTC timestamp
    uint64 timestamp = 2;

    CatalogProductCreateAction catalog_product_create = 3; 
    CatalogProductUpdateAction catalog_product_update = 4; 
    CatalogProductDeleteAction catalog_product_delete = 5; 
} 
```

```
message CatalogProductOperationPayload { 
    enum Actions { 
        UNSET_ACTION = 0;
        SET_CATALOG_PRODUCT_STATUS = 1;
    }

    Action action = 1;

    // Approximately when transaction was submitted, as a Unix UTC timestamp
    uint64 timestamp = 2;

    SetCatalogProductStatusAction set_catalog_product_status = 3;  
} 
```

## Catalog Actions

### CatalogCreateAction
CatalogCreateAction adds a new catalog to state. The transaction should be submitted 
by an agent, which is identified by its signing key, acting on behalf of the 
organization that corresponds to the owner in the create transaction. (Organizations 
and agents are defined by the Pike smart contract.)
```
message CatalogCreateAction { 

>     // GS1 Company Prefix from owner + catalog_id are use as 

Reviewer: s/use/used/

    // a composite key for determining the state address
    string owner = 1;
    string catalog_id = 2;
    string name = 3;
    repeated PropertyValues properties = 4; 
} 
```
Validation requirements:

- If a catalog with catalog_id already exists the transaction is invalid.
- The signer of the transaction must be an agent in the Pike state and must belong 
to an organization in Pike state, otherwise the transaction is invalid.
- The agent must have the permission can_create_catalog for the organization, 
otherwise the transaction is invalid.

> If all requirements are met, the transaction will be accepted, the batch will be 

Reviewer: We shouldn't pull Sawtooth specific details into this description (batches & blocks).

Reviewer: Also repeated below.

Author: ✅

written to a block, and the catalog will be created in state.

The inputs for CatalogCreateAction must include:

> Address of the Agent submitting the transaction

Reviewer 7: This is not formatted correctly (same with the rest of inputs/outputs)

Reviewer 7: This is still an issue

Author: ✅

Author: ✅

Address of the Organization the Catalog is being created for
Address of the Catalog to be created

The outputs for CatalogCreateAction must include:

Address of the Catalog created

### CatalogUpdateAction
CatalogUpdateAction adds a new catalog to state. The transaction should be submitted by 
an agent, which is identified by its signing key, acting on behalf of the organization 

> that corresponds to the owner in the delete transaction. (Organizations and agents are 

Reviewer 7: Should "delete transaction" be update transaction or create transaction?

Author: ✅

defined by the Pike smart contract.)
```
message CatalogUpdateAction { 

>     // GS1 Company Prefix from owner + catalog_id are use as 

Reviewer: s/use/used/

} 
```
Validation requirements:

- If a catalog with catalog_id exists the transaction is invalid.
- The signer of the transaction must be an agent in the Pike state and must belong to an 
organization in Pike state, otherwise the transaction is invalid.
- The agent must have the permission can_update_catalog for the organization, 
otherwise the transaction is invalid.

If all requirements are met, the transaction will be accepted, the batch will be written 
to a block, and the product will be created in state.

The inputs for CatalogUpdateAction must include:

Address of the Agent submitting the transaction
Address of the Organization the Catalog is being created for
Address of the Catalog to be updated

The outputs for CatalogUpdateAction must include:

Address of the Catalog to be updated

### CatalogDeleteAction
CatalogDeleteAction adds a new catalog to state. The transaction should be submitted 
by an agent, which is identified by its signing key, acting on behalf of the 
organization that corresponds to the owner in the delete transaction. 
(Organizations and agents are defined by the Pike smart contract.)
```
message CatalogDeleteAction { 

>     // GS1 Company Prefix from owner + catalog_id are use as 

Reviewer: s/use/used/

the transaction is invalid.

If all requirements are met, the transaction will be accepted, the batch will be 
written to a block, and the catalog will be created in state.

The inputs for CatalogDeleteAction must include:

Address of the Agent submitting the transaction
Address of the Organization the Catalog is being created for
Address of the Catalog to be deleted

The outputs for CatalogDeleteAction must include:

Address of the Catalog to be deleted

## Catalog_Product Actions

### CatalogProductCreateAction
The CatalogProductCreateAction adds a new catalog_product to state. The transaction should 
be submitted by an agent, which is identified by its signing key, acting on behalf of the 
organization that corresponds to the owner in the create transaction. (Organizations and 
agents are defined by the Pike smart contract.)
```
message CatalogProductCreateAction { 
    // catalog_id and product_id are used in deriving the state address
    string catalog_id = 1
    string product_id = 2 // Reference from Grid Product 
    // schema defined additional fields
    repeated PropertyValues properties = 4;
} 
```

> Validation requirements:

Reviewer 7: Does there need to be a product in normal product state that correlates to this CatalogProduct?

Author: Changed to catalog_product_id ✅

- If a catalog_product with product_id and catalog_id already exists the transaction is invalid.
- The signer of the transaction must be an agent in the Pike state and must belong to an 
organization in Pike state, otherwise the transaction is invalid.
- The agent must have the permission can_create_product for the organization, otherwise the 
transaction is invalid.
- If the product_namespace is GS1, the organization must contain a GS1 Company Prefix in its 
metadata (gs1_company_prefixes), and the prefix must match the company prefix in the 
product_id, which is a GTIN if GS1, otherwise the transaction is invalid.

> - The properties must be valid for the catalog_product schema. Meaning its properties must only contain properties that are included in the catalog_product Schema. 

Reviewer: s/schema. Meaning its properties/schema; its properties/

Author: ✅

- Address of the Catalog_Product created

### CatalogProductUpdateAction
CatalogProductUpdateAction updates an existing product in state. The transaction should be submitted 
by an agent, identified by its signing key, acting on behalf of an organization that 
corresponds to the owner in the product being updated. (Organizations and agents are defined 
by the Pike smart contract.)
```
message CatalogProductUpdateAction { 
    // catalog_id and product_id are used in deriving the state address
    string catalog_id = 1; 
    string product_id = 2;
    // this will replace all properties currently defined
    repeated PropertyValues properties = 4; 
} 
```
Validation requirements:

- If a catalog_product with product_id does not exist the transaction is invalid.
- The signer of the transaction must be an agent in the Pike state and must belong to an 
organization in Pike state, otherwise the transaction is invalid.
The owner in the product must match the organization that the agent belongs to,
otherwise the transaction is invalid.
- The agent must have the permission can_update_product for the organization, otherwise 
the transaction is invalid.

> - The properties must be valid for the catalog_product schema. Meaning its properties must only contain properties that are included in the catalog_product Schema. 

Reviewer: Check word wrap.

Author: ✅

The inputs for CatalogProductUpdateAction must include:

- Address of the Agent submitting the transaction
- Address of the Organization the Product is being updated for
- Address of the Product to be updated

> The outputs for ProductUpdateAction must include:

Reviewer 7: ProductUpdateAction -> CatalogProductUpdateAction

Author: ✅

- Address of the updated catalog_product

### CatalogProductDeleteAction
CatalogProductDeleteAction removes an existing product from state. The transaction should be 
submitted by an agent, identified by its signing key, acting on behalf of the organization 

> that corresponds to the org_id in the product being updated. (Organizations and agents are 

Reviewer 7: "product being updated." ->  "product being deleted."

SetCatalogProductStatusAction updates an existing catalog_product in state. The transaction should be submitted 
by an agent, identified by its signing key, acting on behalf of an organization that 
corresponds to the owner in the product being updated. (Organizations and agents are defined 
by the Pike smart contract.)

```
message SetCatalogProductStatusAction { 
    enum Status { 
        INACTIVE = 0;
        ACTIVE = 1; 
        DISCONTINUED = 2;
    }
    // catalog_id and product_id are used in deriving the state address
    string catalog_id = 1; 
    string product_id = 2;
    Status catalog_product_status  = 4; 
    // Reason for the change
    string reason = 5;
} 
```

Validation requirements:

- If a catalog_product with product_id does not exist the transaction is invalid.
- The signer of the transaction must be an agent in the Pike state and must belong to an 
organization in Pike state, otherwise the transaction is invalid.
The owner in the product must match the organization that the agent belongs to,
otherwise the transaction is invalid.
- The agent must have the permission can_update_product for the organization, otherwise 
the transaction is invalid.

> - The properties must be valid for the catalog_product schema. Meaning its properties must only contain properties that are included in the catalog_product Schema. 

Reviewer: s/schema. Meaning its properties/schema; its properties/

Reviewer: Word wrap

Author: ✅

The inputs for SetCatalogProductStatusAction must include:

- Address of the Agent submitting the transaction
- Address of the Organization the catalog_product is being updated for
- Address of the catalog_product to be updated

The outputs for SetCatalogProductStatusAction must include:

- Address of the updated catalog_product

# Rationale and alternatives
[alternatives]: #alternatives

The design of catalog in this manner leaves it extensible for domain specific
use. The concept of price has been discussed and it's been decided to include 
the notion of price within products added to a catalog. A simple implementation 

> of price is fine and currently serves as a placeholder, while a more robust 

Reviewer: s/is fine and//

Author: ✅

"length") with a data type (such as integer).

## Transactions

Catalogs are managed by submitting transactions to Hyperledger Grid, which will
process them with the Grid Catalog smart contract. The following transactions
are supported:

**catalog actions:**

- CatalogCreate - create a catalog and store it in state
- CatalogUpdate - update a catalog in state
- CatalogDelete - deletes a catalog from state and associated catalog_products

**catalog_product actions:**

- CatalogProductCreate - create a catalog_product and stores it in state
- CatalogProductUpdate - updates a catalog_product in state
- CatalogProductDelete - delete a catalog_product from state
- CatalogProductSetStatus - changes the "status" of a catalog_product to an
  active, inactive, or discontinued state

The CatalogProductSetStatus action will accept a parameter list of the
catalogs the action should be performed on. Meaning Activate, Deactivate, or 
Discontinue can be performed on a single, multiple, or all catalogs that an 
organization has.

## Permissions

Creation of a Grid Catalog is restricted to agents acting on behalf of the
organization in the catalog's owner field. An organization must have the GS1

> company prefix in the "gs1_company_prefixes" metadata field for its Pike

Author: added a link ✅

need the operational permission to perform the desired operation on a Grid
Catalog.

```
CatalogProductSetStatus: Agent from org needs the "can_set_catalog_product_status" 
permission
```

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

## State

### Catalog Representation

The primary object stored in state is **Catalog**, which consists of a
**catalog_id** (hash of the catalog_name), an **owner** (org_id
compatible w/Pike), a **Name**, and a repeated field of **PropertyValues**. 
The properties available are defined by the Grid Property Schema Transaction 
Family and are restricted to the fields and rules of the GS1 Catalog schema 
(which will be defined at a later time).  Transactions which are responsible 
for setting catalog state values must ensure that those properties conform with 
the requirements of the GS1 Catalog Schema (not yet defined).

```
message Catalog {
    string catalog_id = 1;
    string owner = 2;
    string name = 3;
    repeated PropertyValue properties = 4;
}
```

### Referencing a Catalog

Catalogs are uniquely referenced by their catalog_id. For example:

```

> get_catalog(catalog_id) // hash(catalog_name)

Reviewer 7: This requires the org_id as well.

Author: added ✅

Reviewer 7: looks like this didnt get pushed

Author: added ✅

catalog_id and product_id. For example:

```
get_catalog_product(catalog_id, product_id) // hash(catalog_name), product_id
set_catalog_product(catalog_id, product_id, catalog_product) //hash(catalog_name), product_id, catalog_product
```

### Catalog Product Addressing in the Merkle-Radix State System

The Grid GS1 Catalog Product state address will be prefixed with the Grid
namespace of "621dee", in addition to "02" (product namespace), "01" for (gs1
product namespace), and the catalog_id.

Therefore, all addresses starting with:

```
"621dee" + "02" + "01 + "cad4e781e797712" // catalog_id
```

are Grid GS1 Catalog Products and are expected to contain the standard
attributes of a Grid GS1 Product as well as the extra PropertyValues that
are defined in the catalog product schema. These PropertyValues are in
addition to the expected PropertyValues of a [Grid GS1
Product](https://github.com/hyperledger/grid-rfcs/blob/fbedec06d70b16492fea9f6b1e87146c5fc56771/0000-product.md).

GTIN formats consist of 14-digit "numeric strings" which include some amount of
internal "0" padding depending on the specific GTIN format (GTIN-12, GTIN-13, or 

> GTIN-14). After the 12-hex-characters that are consumed by the gridnamespace 

Reviewer 7: gridnamespace -> Grid namespace

Author: ✅

> prefix, the product namespace prefix, GS1 namespace prefix, and catalog product 

Reviewer 7: catalog product namespace should be replaced with catalog id and 12 should be to include the number of hex in "621dee" + "02" + "01 + "cad4e781e797712". (or just remove catalog product namespace  and say it is 10 hex and explain adding the catalog id after)

Author: Fixed, also adjusted the number ranges for the remaining characters in the address. There were slightly off. ✅

namespace prefix, there are 58 hex characters remaining in the address.

The 12 to 14 digits of the GTIN can be left padded with 40 to 42-hex-character
zeroes and right padded with 2-hex-character zeroes to accommodate potential
future storage associated with the GS1 Product representation, for example:

```
"621dee" + "02" + "01" + "cad4e781e797712" + "00000000000000000000000000000"
+ 14-character "numeric string" product_id + "00" // product_id == GTIN
```

A full Catalog Product address using the example GTIN from
https://www.gtin.info/ would therefore be:

```
"621dee0201cad4e781e797712000000000000000000000000000000001234560001200"
```

### Catalog Product Schema Definition

This schema defines the additional attributes required of a grid product, such
that it may be considered a "catalog product". Catalog products should be 
grouped with other catalog products. Once grouped, the grid catalog can be 
shared with other participants in a grid network.

Any grid product added to a catalog (thus becoming a catalog product) will need
to contain a reference fields to the original product (product_id, 

> product_namespace,and owner) and a catalog_id. A catalog product schema may 

Reviewer 7: missing space between ",and"

Author: ✅

> also be defined torequire additional fields for that catalog_product. This 

Reviewer 7: to require -> to require

Author: ✅

example schema enforces a _required_ **_catalog_id_** & **_status_** field, as 
well as an _optional_ **_prices_** & **_return_policy_** field.

Example schema:

```
Schema(
    name="Catalog Product",
    description="A grid product that can be grouped with other products via
grid catalog",
    owner = "Target"
    properties=[
        PropertyDefinition(
            name="catalog_id",
            data_type=PropertyDefinition.DataType.STRING,
            description="The id of the catalog this catalog product belongs to"
            required=True
        ),
        PropertyDefinition(
            name="status",
            data_type=PropertyDefinition.DataType.ENUM,
            description="The current state of the catalog product",

>             enum_options=["ACTIVE", "INACTIVE"],

Reviewer 7: Should this include discontinued?

Author: Yes, added ✅

    owner='Target', // Reference from Grid Product
    properties=[
        PropertyValue(
            name="catalog_id",
            data_type=PropertyDefinition.DataType.STRING,
            string_value="e70fe351d1d096c"
        ),
        PropertyValue(
            name="status",
            data_type=PropertyDefinition.DataType.ENUM,
            enum_value=1, # ACTIVE (default)
        ),
        PropertyValue(
            name="price"
            data_type=PropertyDefinition.DataType.STRING,
            string_value="1-10:$50, 10-50:$40, 50+:$30"  # price windows
        ),
        PropertyValue(
            name="return policy"
            data_type=PropertyDefinition.DataType.STRING,
            string_value="30 days from sale"
        ),
        # ... Additional Product Properties
    ])
```

A catalog product, from a data model perspective, is a grid product. With the 
differentiators being:
- It contains the additional fields defined by the catalog_product schema
- It's stored in the catalog_product state location within the merkle tree

## Transaction Payload and Execution

The following payload contains an action enum and the associated action

> payload.This allows for the action payload to be dispatched to the appropriate 

Reviewer 7: payload.This ->  payload. This

Author: ✅

logic. Only the defined actions (enum) are available, and only one Action 
should be performed in a payload.

### CatalogPayload Transaction

```
message CatalogPayload {
    enum Actions {
        UNSET_ACTION = 0;
        CATALOG_CREATE = 1;
        CATALOG_UPDATE = 2;
        CATALOG_DELETE = 3;
        CATALOG_PRODUCT_CREATE = 100;
        CATALOG_PRODUCT_UPDATE = 101;
        CATALOG_PRODUCT_DELETE = 102;
        CATALOG_PRODUCT_SET_STATUS = 103;
    }

    Action action = 1;

    // Approximately when transaction was submitted, as a Unix UTC timestamp
    uint64 timestamp = 2;

    CatalogCreateAction catalog_create = 3;
    CatalogUpdateAction catalog_update = 4;
    CatalogDeleteAction catalog_delete = 5;

>     CatalogProductCreateAction catalog_product_create = 6;

Reviewer 7: This should jump in number along with the enum, start at 100 or something.

Author: Are you suggesting something like this?

```
message CatalogPayload {
    enum Actions {
        UNSET_ACTION = 0;
        CATALOG_CREATE = 1;
        CATALOG_UPDATE = 2;
        CATALOG_DELETE = 3;
        CATALOG_PRODUCT_CREATE = 100;
        CATALOG_PRODUCT_UPDATE = 101;
        CATALOG_PRODUCT_DELETE = 102;
        CATALOG_PRODUCT_SET_STATUS = 103;
    }

    Action action = 1;

    // Approximately when transaction was submitted, as a Unix UTC timestamp
    uint64 timestamp = 2;

    CatalogCreateAction catalog_create = 101;
    CatalogUpdateAction catalog_update = 102;
    CatalogDeleteAction catalog_delete = 103;
    CatalogProductCreateAction catalog_product_create = 104;
    CatalogProductUpdateAction catalog_product_update = 105;
    CatalogProductDeleteAction catalog_product_delete = 106;
    CatalogProductSetStatusAction set_catalog_product_status = 107;
}
```

Reviewer 7: Yes this is what i meant

submitted by an agent, which is identified by its signing key, acting on behalf 
of the organization that corresponds to the owner in the create transaction. 
(Organizations and agents are defined by the Pike smart contract.)

```
message CatalogCreateAction {
    // GS1 Company Prefix from owner + catalog_id are use as
    // a composite key for determining the state address
    string owner = 1;
    string catalog_id = 2;
    string name = 3;
    repeated PropertyValues properties = 4;
}
```

Validation requirements:

- If a catalog with catalog_id already exists the transaction is invalid.
- The signer of the transaction must be an agent in the Pike state and must
  belong to an organization in Pike state, otherwise the transaction is invalid.
- The agent must have the permission can_create_catalog for the organization,
  otherwise the transaction is invalid.

The inputs for CatalogCreateAction must include:

Address of the Agent submitting the transaction
Address of the Organization the Catalog is being created for
Address of the Catalog to be created

The outputs for CatalogCreateAction must include:

Address of the Catalog created

### CatalogUpdateAction

> CatalogUpdateAction adds a new catalog to state. The transaction should be

Reviewer 7: Should this say updates an existing catalog in state.

Author: ✅

Reviewer 7: This still says "new catalog" should say existing

Author: fixed ✅

submitted by an agent, which is identified by its signing key, acting on behalf 
of the organization that corresponds to the owner in the update transaction. 
(Organizations and agents are defined by the Pike smart contract.)

```
message CatalogUpdateAction {
    // GS1 Company Prefix from owner + catalog_id are use as
    // a composite key for determining the state address
    string owner = 1;
    string catalog_id = 2;
    string name = 3;
    repeated PropertyValues properties = 4;
}
```

Validation requirements:

- If a catalog with catalog_id exists the transaction is invalid.
- The signer of the transaction must be an agent in the Pike state and must
  belong to an organization in Pike state, otherwise the transaction is invalid.
- The agent must have the permission can_update_catalog for the organization,
  otherwise the transaction is invalid.

The inputs for CatalogUpdateAction must include:

> Address of the Agent submitting the transaction

Author: formatted in the sense of? wrapping them as code blocks?

> Address of the Agent submitting the transaction

Author: Oh adding them as bullets. ✅

Address of the Organization the Catalog is being created for
Address of the Catalog to be updated

The outputs for CatalogUpdateAction must include:

Address of the Catalog to be updated

### CatalogDeleteAction

> CatalogDeleteAction adds a new catalog to state. The transaction should be

Reviewer 7: deletes a catalog

Author: ✅

Reviewer 7: It now says "deletes a new catalog to state" should say "deletes an existing catalog to state"

Author: added ✅

submitted by an agent, which is identified by its signing key, acting on behalf 
of the organization that corresponds to the owner in the delete transaction.
(Organizations and agents are defined by the Pike smart contract.)

```
message CatalogDeleteAction {
    // GS1 Company Prefix from owner + catalog_id are use as
    // a composite key for determining the state address
    string owner = 1;
    string catalog_id = 2;
}
```

Validation requirements:

- If a catalog with catalog_id exists the transaction is valid, otherwise it's
  invalid.
- The signer of the transaction must be an agent in the Pike state and must
  belong to an organization in Pike state, otherwise the transaction is invalid.
- The agent must have the permission can_delete_catalog for the organization,
  otherwise the transaction is invalid.

The inputs for CatalogDeleteAction must include:

> Address of the Agent submitting the transaction

Author: ✅

of master data elements. Using the Grid Catalog, organizations can share
collections of related/unrelated products with other participants in their
network.

# Guide-level explanation

[guide-level-explanation]: #guide-level-explanation

## Entities

A **catalog** is a high-level construct that **catalog_products** can
reference. A _catalog_product_ is an instance of a Grid Product with additional 
schema defined attributes specific to the trading partner involved. The 
"catalog_product" schema can be extended to include any number of attributes that 
are required by an organization. A catalog can be referenced, or used to share 
item level information in a supply chain. A catalog is referenced by a 
"catalog_product" via its **catalog_id**. A catalog also has an **owner** (the 
organization that creates the catalog). Also a **Name** for the catalog.

A catalog, like a product, can also have one or more **properties**. Properties
are described in the Grid Primitives RFC. A _property namespace_ contains
multiple _property schemas_. A property schema associates a name (such as
"length") with a data type (such as integer). A catalog may want additional 

> properties to support execute custom smart contract, or trigger some type of 

Reviewer 7: or "to support the execution of"

> properties to support execute custom smart contract, or trigger some type of 

Author: ✅

check. An example could be an expiry_date for an entire catalog. 

## Transactions

Catalogs are managed by submitting transactions to Hyperledger Grid, which will
process them with the Grid Catalog smart contract. The following transactions
are supported:

**catalog actions:**

- CatalogCreate - create a catalog and store it in state
- CatalogUpdate - update a catalog in state
- CatalogDelete - deletes a catalog from state and associated catalog_products

**catalog_product actions:**

- CatalogProductCreate - create a catalog_product and stores it in state
- CatalogProductUpdate - updates a catalog_product in state
- CatalogProductDelete - delete a catalog_product from state
- CatalogProductSetStatus - changes the "status" of a catalog_product to an
  active, inactive, or discontinued state

The CatalogProductSetStatus action will accept a parameter list of the

> catalogs the action should be performed on. Meaning Activate, Deactivate, or 

Author: Changed to a repeated field: 
```
message CatalogProductSetStatusAction {
    enum Status {
        INACTIVE = 0;
        ACTIVE = 1;
        DISCONTINUED = 2;
    }
    // catalog_id and product_id are used in deriving the state address
    repeated catalog_id = 1;
    string product_id = 2;
    Status catalog_product_status  = 4;
    // Reason for the change
    string reason = 5;
}
```

I'm thinking the implementation of the action can handle applying the status update across the specified catalogs.

information can be found [here](https://sawtooth.hyperledger.org/docs/core/releases/1.0/transaction_family_specifications/settings_transaction_family.html).

Validation requirements:

- If a catalog_product with catalog_product_id does not exist the transaction is
  invalid.
- The signer of the transaction must be an agent in the Pike state and must
  belong to an organization in Pike state, otherwise the transaction is invalid.
- The owner in the product must match the organization that the agent belongs
  to, otherwise the transaction is invalid.
- The agent must have the permission “can_delete_product” for the organization
  otherwise the transaction is invalid.

The inputs for CatalogProductDeleteAction must include:

- Address of the Agent submitting the transaction
- Address of the Organization the Product is being deleted for
- Address of the catalog_product to be deleted

The outputs for CatalogProductDeleteAction must include:

- Address of the catalog_product to be deleted

**_NOTE: Deleting a catalog_product is potentially dangerous operation that
could leave dangling references and should be done with care._**

## Catalog_Product Operations

> ### CatalogProductSetStatus

Reviewer 7: This section should be updated to explain that it can update the the status of multiple Catalog Products. For example it should explain that every product must be able to be updated or the transaction is invalid,  and change inputs and outputs to be for the multiple products and so on.

Author: ✅

Author: ✅

Reviewer 3: I might be missing the update for the above comment but I don't see the RFC contents reflect the request. I have a follow on to the item @agunde406 brings up and that is:

```
if an attempt to update multiple Catalog Products in 
one transaction is requested, should the RFC reflect
the expectation to apply the outcome at the whole 
transaction level or it is acceptable to handle each Product 
individually?
```

- Feature Name: Product Catalog
- Start Date: 2019-07-12
- RFC PR:
- Hyperledger Grid Issue: [GRID-95](https://jira.hyperledger.org/browse/GRID-95)

# Summary

[summary]: #summary

This RFC proposes a generic implementation for a Hyperledger Grid _Product
Catalog_. Product Catalogs will be represented by a unique identifier, which
will be referenced by the _catalog_products_ included in the catalog. That
catalog_id can be used to share a grouping of products that can be shared with 
one or more organizations. The Grid Catalog will contain 4 fields. A 

> **catalog_id**, a **catalog_owner**, a **Name** for the catalog, and a repeated 

Reviewer 9: I noticed some inconsistencies with the capitalization and text emphasis of titles, messages, and field names throughout the RFC. 

For example, I think **Name** be lowercase here. Also, in the next paragraph down they have an italics emphasis applied. I'd probably recommend using the `code` emphasis instead for these to make it clear that you're referring to actual fields in the messages.

I think a quick pass through to clean those up would make the RFC a lot more clear!

Author: ✅  updated, let me know if I missed anything

> field of PropertyValues for custom fields.

Reviewer 3: Consider naming `PropertyValue` to `PropertyDefinition`

Author: `PropertyDefinition` is the term used to define a `PropertyValue` in the context of a Grid Schema. You can toss in `PropertyValues` without the need for a schema. 

Though if you wanted one or more Catalogs to be _required_ to have certain additional fields you'd need to write a Grid Schema that has `PropertyDefinitions` for the `PropertyValues` you wish to collect.

catalog will adhere to the "catalog_product" schema (or another custom-defined 
schema). The example schema in particular will enforce a required `catalog_id`
field, a required `status` field, an optional `return_policy`, and an 
optional `price` field.

# Motivation

[motivation]: #motivation

The Grid Catalog implementation is designed for sharing an assortment of
products between organizations. Each product in a catalog will contain a number
of master data elements. Using the Grid Catalog, organizations can share
collections of related/unrelated products with other participants in their
network.

# Guide-level explanation

[guide-level-explanation]: #guide-level-explanation

## Entities

A `catalog` is a high-level construct that `catalog_products` can
reference. A _catalog_product_ is an instance of a Grid Product with additional 
schema defined attributes specific to the trading partner involved. The 
"catalog_product" schema can be extended to include any number of attributes that 
are required by an organization. A catalog can be referenced, or used to share 
item level information in a supply chain. A catalog is referenced by a 

> "catalog_product" via its `catalog_id`. A catalog also has an `owner` (the 

Reviewer 3: Above, `owner` is called `catalog_owner`. If these are pointing to the same parameter, the name should be consistent.

Author: ✅

organization that creates the catalog). Also a `name` for the catalog.

A catalog, like a product, can also have one or more `properties`. Properties
are described in the Grid Primitives RFC. A _property namespace_ contains
multiple _property schemas_. A property schema associates a name (such as
"length") with a data type (such as integer). A catalog may want additional 
properties to support executing custom smart contract, or trigger some type of 
check. An example could be an expiry_date for an entire catalog. 

## Transactions

Catalogs are managed by submitting transactions to Hyperledger Grid, which will
process them with the Grid Catalog smart contract. The following transactions
are supported:

**catalog actions:**

- CatalogCreate - create a catalog and store it in state
- CatalogUpdate - update a catalog in state
- CatalogDelete - deletes a catalog from state and associated catalog_products

> 

Reviewer 3: Would `CatalogRead` be valid?

Author: I think a CatalogRead event could be handled at an application level. Would there be a compelling reason to include it in the protocol? Distributing the information to other nodes that someone has read something, keeping track of who has read what and when? 

Possibly, again it could also be handled at an application level.

Reviewer 3: Chatted with James and Shawn on some additional detail. I agree with your response 👍

**catalog_product actions:**

> - CatalogProductCreate - create a catalog_product and stores it in state

Reviewer: s/create/creates/

Author: ✅

CatalogProductDelete: Agent from org needs the "can_delete_product" permission
```

In the case of **catalog operations** agents from the org they represent will
need the operational permission to perform the desired operation on a Grid
Catalog.

```
CatalogProductSetStatus: Agent from org needs the "can_set_catalog_product_status" 
permission
```

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

## State

### Catalog Representation

The primary object stored in state is `Catalog`, which consists of a
`catalog_id` (hash of the catalog_name), an `owner` (org_id
compatible w/Pike), a `name`, and a repeated field of `PropertyValues`. 
The properties available are defined by the Grid Property Schema Transaction 
Family and are restricted to the fields and rules of the GS1 Catalog schema 
(which will be defined at a later time).  Transactions which are responsible 
for setting catalog state values must ensure that those properties conform with 
the requirements of the GS1 Catalog Schema (not yet defined).

```
message Catalog {
    string catalog_id = 1;
    string owner = 2;
    string name = 3;

>     repeated PropertyValue properties = 4;

Reviewer 3: Consider `PropertyDefinition` ... ?

Author: See explanation above

}
```

### Referencing a Catalog

Catalogs are uniquely referenced by their catalog_id. For example:

```
get_catalog(org_id, catalog_id) // GS1 Company Prefix, hash(catalog_name)
set_catalog(org_id, catalog_id, catalog) // GS1 Company Prefix, hash(catalog_name), catalog
```

### Catalog Addressing in the Merkle-Radix State System

In order to uniquely locate Catalogs in the Merkle-Radix state system, an
address must be constructed which identifies the storage location of the Catalog
representation.

All Grid addresses are prefixed by the 6-hex-character namespace prefix
"621dee", Catalogs are further prefixed under the Grid namespace with reserved
enumerations of "03" ("00", "01", and "02" being reserved for other purposes)
indicating "Grid Catalogs" and an additional org_id. Currently in Grid Product,
org_ids are a [6-10 digit GS1 company prefix](https://www.gs1-us.info/gs1-company-prefix/).
To keep the prefix length consistent, any GS1 company prefix less than 10
digits will be left padded with the hexadecimal representation of the number 10 

> ("a") until it is 10 digits in length. This composite address will enable 

Author: There is a possibility of the numeric GS1 company prefix starting with a "0". The reason "a" was chosen was to be able to tell where the variable-length GS1 company prefix starts/how long it is.

> ("a") until it is 10 digits in length. This composite address will enable 

Reviewer 3: I recommend adding that response into the readme directly.

inexpensive catalog operations.

Therefore, all addresses starting with:

```
"621dee" + "03" + "aaaa123456" // org_id (in this case a gs1_company_prefix)
```

are Grid GS1 Catalogs identified by the hash of the catalog name which can be
referenced by a "catalog_product" to group products to a specific catalog.

> The catalog_id format consists of 70-digit "alphanumeric string" which include

Reviewer 3: I think I understand what this is saying but the wording reads weird. Open to suggestions?

Author: Changed to:

> The catalog_id is a 70-digit "alphanumeric string" which includes
a fixed amount of internal "0" padding.

a little less wordy

> The catalog_id format consists of 70-digit "alphanumeric string" which include

Reviewer: s/consists of 70-digit/consists of a 70-digit/

Also, this should probably read 70-character, instead of 70-digit.

Author: ✅

associated with the GS1 Catalog representation, for example:

```
"621dee" + "03" + "aaaa123456" + "00000000000000000000000000000000000" + catalog_id 
+ "00" // catalog_id == hash(catalog_name).trunc(15)
```

Using SHA-2 from the rust-crypto crate we can tie the catalog_id to catalog_name
to prevent duplication of catalogs when invoking the catalog_create action.

```
extern crate crypto;
use self::crypto::digest::Digest;
use crypto::sha2::Sha512;

fn sha2_512() {
    // create a SHA2-512 hasher
    let mut hasher = Sha512::new();

    // write input message
    hasher.input_str("catalog_name");

    // read hash digest
    let hex = hasher.result_str();
    let hex_15 = &hex[0..15];

    println!("SHA2-512 Hash is {:?}", hex);
    println!("SHA2-512 Hash (15) {:?}", hex_15);
}

>> SHA2-512 Hash is:
cad4e781e7977123893a42c7e7230e798bf5693f95c04980ba70bbe3504b194227c62e7e32d83ddb
d083979720d7025b2cc9f68ac7b572dc6a607409b2dcdc43
>> SHA2-512 Hash (15): cad4e781e797712
```

The full catalog address would look like:

```
"621dee03aaaa12345600000000000000000000000000000000000cad4e781e79771200"
```

### Referencing a Catalog Product

> Catalogs products are uniquely referenced by a composite key composed of their

Reviewer: s/Catalogs products/Catalog Products/

Author: ✅

addition to the expected PropertyValues of a [Grid GS1
Product](https://github.com/hyperledger/grid-rfcs/blob/fbedec06d70b16492fea9f6b1e87146c5fc56771/0000-product.md).

GTIN formats consist of 14-digit "numeric strings" which include some amount of
internal "0" padding depending on the specific GTIN format (GTIN-12, GTIN-13, or 
GTIN-14). After the 10-hex-characters that are consumed by the Grid namespace 
prefix, the product namespace prefix, and GS1 namespace prefix. There are 
15-hex-characters consumed by the catalog id, leaving 45 hex characters
remaining in the address.

The 11 to 13 digits of the GTIN can be left padded with 30 to 32-hex-character
zeroes and right padded with 2-hex-character zeroes to accommodate potential
future storage associated with the GS1 Product representation, for example:

```
"621dee" + "02" + "01" + "cad4e781e797712" + "00000000000000000000000000000"
+ 14-character "numeric string" product_id + "00" // product_id == GTIN
```

A full Catalog Product address using the example GTIN from
https://www.gtin.info/ would therefore be:

```
"621dee0201cad4e781e797712000000000000000000000000000000001234560001200"
```

### Catalog Product Schema Definition

> This schema defines the additional attributes required of a grid product, such

Reviewer: s/grid/Grid/

Author: ✅

that it may be considered a "catalog product". Catalog products should be 

> grouped with other catalog products. Once grouped, the grid catalog can be 

Reviewer 3: Should `grid` be `Grid`?

Author: Updated capitalization across the rfc, good callout

shared with other participants in a grid network.

Any grid product added to a catalog (thus becoming a catalog product) will need
to contain a reference fields to the original product (product_id, 
product_namespace, and owner) and a catalog_id. A catalog product schema may 
also be defined to require additional fields for that catalog_product. This 
example schema enforces a _required_ `catalog_id` & `status` field, as 
well as _optional_ `prices` & `return_policy` field.

Example schema:

```
Schema(
    name="Catalog Product",
    description="A grid product that can be grouped with other products via
grid catalog",

>     owner = "Target"

Reviewer 3: xref above question on the naming of this attribute. Should it be `catalog_owner`?

Author: I changed the above to just `owner`. Though, the reference above was to the owner of a catalog (the entity that created it). This owner is the entity that creates a `catalog_product` by extending the definition of its existing Grid product.

            description="The price of the catalog product"
            required=False
        ),
        PropertyDefinition(
            name="return_policy",
            data_type=PropertyDefinition.DataType.STRING,
            description="The return policy of the of the catalog product"
            required=False
        )])
```

In order to store a catalog product in global state, we will extend from the
definition of a grid product with several fixed fields, namely identifiers, as
well as the set of dynamic property values that will conform to the catalog
product schema definition.

**Grid Product from the [Product
RFC](https://github.com/hyperledger/grid-rfcs/blob/master/text/0005-product.md):**

```
message Product {
    enum ProductNamespace {
        UNSET_NAMESPACE = 0;
        GS1 = 1;
    }

    ProductNamespace product_namespace = 1;
    string product_id = 2;
    string owner = 3;
    repeated PropertyValue properties = 4;
}
```

The catalog_product smart contract would then be responsible for validating the
properties field against the CatalogProduct schema at run-time.

A catalog_product entry would reference Grid Product. In addition to

> the properties defined in the CatalogProduct schema. The data model

Author: yes, ✅

    }

    Action action = 1;

    // Approximately when transaction was submitted, as a Unix UTC timestamp
    uint64 timestamp = 2;

    CatalogCreateAction catalog_create = 3;
    CatalogUpdateAction catalog_update = 4;
    CatalogDeleteAction catalog_delete = 5;
    CatalogProductCreateAction catalog_product_create = 100;
    CatalogProductUpdateAction catalog_product_update = 101;
    CatalogProductDeleteAction catalog_product_delete = 102;
    CatalogProductSetStatusAction set_catalog_product_status = 103;
}
```

## Catalog Actions

### CatalogCreateAction

CatalogCreateAction adds a new catalog to state. The transaction should be
submitted by an agent, which is identified by its signing key, acting on behalf 
of the organization that corresponds to the owner in the create transaction. 
(Organizations and agents are defined by the Pike smart contract.)

```
message CatalogCreateAction {
    // GS1 Company Prefix from owner + catalog_id are use as
    // a composite key for determining the state address
    string owner = 1;
    string catalog_id = 2;
    string name = 3;
    repeated PropertyValues properties = 4;
}
```

Validation requirements:

> - If a catalog with catalog_id already exists the transaction is invalid.

Reviewer: s/already exists the/already exists, the/

Author: ✅

- The signer of the transaction must be an agent in the Pike state and must
  belong to an organization in Pike state, otherwise the transaction is invalid.
- The agent must have the permission can_create_catalog for the organization,
  otherwise the transaction is invalid.

The inputs for CatalogCreateAction must include:

- Address of the Agent submitting the transaction
- Address of the Organization the Catalog is being created for
- Address of the Catalog to be created

The outputs for CatalogCreateAction must include:

Address of the Catalog created

### CatalogUpdateAction

CatalogUpdateAction updates an existing catalog in state. The transaction should be
submitted by an agent, which is identified by its signing key, acting on behalf 
of the organization that corresponds to the owner in the update transaction. 
(Organizations and agents are defined by the Pike smart contract.)

```
message CatalogUpdateAction {
    // GS1 Company Prefix from owner + catalog_id are use as
    // a composite key for determining the state address
    string owner = 1;
    string catalog_id = 2;
    string name = 3;
    repeated PropertyValues properties = 4;
}
```

Validation requirements:

> - If a catalog with catalog_id exists the transaction is invalid.

Reviewer 3: I don't think this is what you want ... if the `catalog_id` does not exist, how would the entity be updated?

Author: You are correct... typo, thanks! ✅

- Address of the catalog_product to be deleted

**_NOTE: Deleting a catalog_product is potentially dangerous operation that
could leave dangling references and should be done with care._**

## Catalog_Product Operations

### CatalogProductSetStatus

CatalogProductSetStatusAction updates an existing catalog_product in state. The
transaction should be submitted by an agent, identified by its signing key, 
acting on behalf of an organization that corresponds to the owner in the 
product being updated. (Organizations and agents are defined by the Pike smart 
contract.)

The implementation of the CatalogProductSetStatusAction will handle iterating 
through the list of catalog_ids, and update the status of the 
catalog_product(s) accordingly. Having the address of an individual 
catalog_product be a composite key containing the catalog_id and product_id 
enables us to easily update the status across one, all, or specific catalogs.

```
message CatalogProductSetStatusAction {
    enum Status {
        INACTIVE = 0;
        ACTIVE = 1;
        DISCONTINUED = 2;
    }
    // catalog_id and product_id are used in deriving the state address

>     repeated catalog_id = 1;

Reviewer 7: `repeated catalog_id = 1;`

>     repeated string catalog_id = 1;

Author: ✅

operation and need to be done with care._

To perform any of the **catalog actions** agents will need the following
permissions:

```
CatalogCreate: Agent from org needs the "can_create_catalog" permission
CatalogUpdate: Agent from org needs the "can_update_catalog" permission
CatalogDelete: Agent from org needs the "can_delete_catalog" permission
```

To perform any of the **catalog_product actions** agents inherit the same
permissions needed for a Grid Product:

```
CatalogProductCreate: Agent from org needs the "can_create_product" permission
CatalogProductUpdate: Agent from org needs the "can_update_product" permission
CatalogProductDelete: Agent from org needs the "can_delete_product" permission
```

In the case of **catalog operations** agents from the org they represent will
need the operational permission to perform the desired operation on a Grid
Catalog.

```
CatalogProductSetStatus: Agent from org needs the "can_set_catalog_product_status" 
permission
```

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

## State

### Catalog Representation

The primary object stored in state is `Catalog`, which consists of a

> `catalog_id` (hash of the catalog_name), an `owner` (org_id

Reviewer 9: This is actually a hash of the catalog's `name`, right?

Author: Correct, the catalog_id is simply a hash of the catalog_name. Collisions in the merkle state are avoided due to the merkle address useing the unique `org_id` as part of its address. It makes it inexpensive to pull all the catalogs associated with a particular org.

Reviewer 9: Oh okay. The `Catalog` message has a field called `name`, but no field called `catalog_name`, which was a bit confusing. I'd recommend updating those field names to match.

Author: Updated

compatible w/Pike), a `name`, and a repeated field of `PropertyValues`. 
The properties available are defined by the Grid Property Schema Transaction 
Family and are restricted to the fields and rules of the GS1 Catalog schema 
(which will be defined at a later time).  Transactions which are responsible 
for setting catalog state values must ensure that those properties conform with 

> the requirements of the GS1 Catalog Schema (not yet defined).

Reviewer 9: Does this mean that actually implementing catalog is dependent on defining the GS1 Catalog Schema?

Author: It's not. This paragraph is a bit redundant and confusing.

Updated:
```
The properties available are defined by the Grid Property Schema Transaction 
Family and are restricted to the fields and rules of the GS1 Catalog schema
(which will be defined at a later time). Currently, you can proceed to create 
a catalog that does not adhere to any schema. The GS1 Catalog Schema will be 
defined in a future RFC.```

```
message Catalog {
    string catalog_id = 1;
    string owner = 2;
    string name = 3;

>     repeated PropertyValue properties = 4;

Reviewer 9: So these properties are extended to any Product that is part of the Catalog. Is that correct?

Author: What do you mean?

The properties here are related to the catalog object itself, and not any products associated with the catalog. 

An example could be adding an "expiry_date" field to a catalog or an "issued_date" field.

  invalid.
- The signer of the transaction must be an agent in the Pike state and must
  belong to an organization in Pike state, otherwise the transaction is invalid.
- The agent must have the permission can_delete_catalog for the organization,
  otherwise the transaction is invalid.

The inputs for CatalogDeleteAction must include:

- Address of the Agent submitting the transaction
- Address of the Organization the Catalog is being created for
- Address of the Catalog to be deleted

The outputs for CatalogDeleteAction must include:

Address of the Catalog to be deleted

**_NOTE: Deleting a catalog is potentially dangerous operation that could leave 
dangling references and should be done with care._**

## Catalog_Product Actions

### CatalogProductCreateAction

The CatalogProductCreateAction adds a new catalog_product to state. The 
catalog_product references a Grid Product for the shared item level master 
data. The transaction should be submitted by an agent, which is identified by 
its signing key, acting on behalf of the organization that corresponds to the
owner in the create transaction. (Organizations and agents are defined by the 
Pike smart contract.)

```

> message CatalogProductCreateAction {

Reviewer 9: So does this create a CatalogProduct object in state, or does it update the Grid Product with the given properties?

Author: It creates a separate object in state known as `catalog_product`. I look at grid product as the item master data about a particular product. The catalog_product derives from that and adds any attributes that are business partner-specific. 

Return_policy is a good example. The Grid Product would likely not have a return_policy field, if it did it would likely be overridden. The catalog_product would contain the return_policy with the specific business partner relationship policy. Does that make sense?

Reviewer 9: Okay, yes that makes sense to me now. Thanks!

The inputs for CatalogProductCreateAction must include:

- Address of the Agent submitting the transaction
- Address of the Organization the Product is being created for
- Address of the Catalog the catalog_product is related too
- Address of the catalog_product namespace schema and the catalog_product’s
  properties must match
- Address of the Grid Product the catalog_product is referencing

The outputs for CatalogProductCreateAction must include:

- Address of the Catalog_Product created

### CatalogProductUpdateAction

CatalogProductUpdateAction updates an existing product in state. The
transaction should be submitted by an agent, identified by its signing key, 
acting on behalf of an organization that corresponds to the owner in the 
product being updated. (Organizations and agents are defined by the Pike 
smart contract.)

```
message CatalogProductUpdateAction {
    // catalog_id and product_id are used in deriving the state address
    string catalog_id = 1;
    string product_id = 2;
    // this will replace all properties currently defined
    repeated PropertyValues properties = 4;
}
```

Validation requirements:

> - If a catalog_product with catalog_product_id does not exist the transaction

Reviewer: s/does not exist the/does not exist, the/

Author: ✅

  otherwise the transaction is invalid.
- The agent must have the permission can_update_product for the organization,
  otherwise the transaction is invalid.
- The properties must be valid for the catalog_product schema; its properties
  must only contain properties that are included in the catalog_product Schema.

The inputs for CatalogProductUpdateAction must include:

- Address of the Agent submitting the transaction
- Address of the Organization the Product is being updated for
- Address of the Product to be updated

The outputs for CatalogProductUpdateAction must include:

- Address of the updated catalog_product

### CatalogProductDeleteAction

CatalogProductDeleteAction removes an existing catalog_product from state. The
transaction should be submitted by an agent, identified by its signing key, 
acting on behalf of the organization that corresponds to the org_id in the 
product being deleted. (Organizations and agents are defined by the Pike smart 
contract.)

```
message CatalogProductDeleteAction {
    // catalog_id and product_id are used in deriving the state address
    string catalog_id = 1;
    string product_id = 2;
}
```

> If the Grid setting Grid.product.allow_delete is set to false, this transaction

Reviewer: s/Grid.product.allow_delete/'grid.product.allow_delete'/

Author: ✅

is invalid (sys admin setting). The default value for `grid.product.allow_delete`
is true. This setting is stored using the Sawtooth Settings smart contract, more 
information can be found [here](https://sawtooth.hyperledger.org/docs/core/releases/1.0/transaction_family_specifications/settings_transaction_family.html).

Validation requirements:

> - If a catalog_product with catalog_product_id does not exist the transaction is

Reviewer: s/does not exist the/does not exist, the/

Author: ✅

acting on behalf of an organization that corresponds to the owner in the 
product being updated. (Organizations and agents are defined by the Pike smart 
contract.)

The implementation of the CatalogProductSetStatusAction will handle iterating 
through the list of catalog_ids, and update the status of the 
catalog_product(s) accordingly. Having the address of an individual 
catalog_product be a composite key containing the catalog_id and product_id 
enables us to easily update the status across one, all, or specific catalogs.

CatalogProductSetStatusAction does not set the status for multiple 
catalog_products. It will only change the status of a single 
catalog_product. The nuance being that this status change can be reflected in 
as many, or as few catalogs an entity desires.

This flexibility is to enable support for use cases like:

- I am discontinuing a catalog_product for entity A, B, C, but not for retailer D.
- I want to discontinue catalog_products for all entity A, B, C, and D.
- I am marking a catalog_product as active for entity A and B, but not C or D just yet.
- I am making an active catalog_product inactive for entity A and B, but not C or D.

```

> message CatalogProductSetStatusAction {

Reviewer 9: So this transaction operates on a catalog product schema. Couldn't there be a case where one of the specified catalog product schemas don't have  the `status` or `status_change_reason` PropertyValues? What would happen in that case?

Also, couldn't you just accomplish this by using the `CatalogProductUpdateAction` transaction, which doesn't make any assumptions about the `PropertyValues`?

Author: The discussion that led to including `status` as a required field of a catalog_product schema was to show that Grid has some opinion of what should be included in a catalog_product. Since catalog_products are intended to be used for business partner-specific relations, we landed on a `status` field (indicating a catalog_products availability) that is something to be shared among all catalog_products.

From an implementation standpoint, this could certainly be implemented as a `CatalogProductUpdateAction`. As I think more about it, I don't see too strong a reason to have this as its own transaction. I'm trying to recall the reasoning of implementing it as a separate transaction, maybe be to enforce the significance of the `status` (availability) field. 

Do you recall the reasoning @davececchi @genasmorgan? You'd be able to explicitly see this transaction on a protocol level, not what the value prop was of that. If there is no strong opposition I'm open to scraping the `CatalogProductSetStatusAction` and just batching it in with `CatalogProductUpdateAction`.

Reviewer 9: If Grid is opinionated about what is in the Catalog Product data model, I would recommend making that its own protobuf and including the status as a field. Otherwise, I don't think it would be ideal to have transactions that explicitly reference specific data that isn't included in the message.

Author: Good point. I think we want to be opinionated, but not forceful. There will be a schema example with status (among other fields) coming out of this RFC. That gives developers a base schema to start with but doesn't force them to use any specific fields. That being said, i'll remove the`CatalogProductSetStatusAction` and that will get batched in with `CatalogProductUpdateAction`

Author: Actually, as I was going to remove the `CatalogProductSetStatusAction` it became clear to me why we needed the `CatalogProductSetStatusAction`. 

In order to set the status of a catalog_product via the CatalogProductUpdateAction you'd need to overwrite all the properties for all the additional fields of that specific catalog, also to do this across more than one catalog, you'd need a separate transaction for each catalog_product. So If I were to discontinue a catalog_product, across 100 catalogs that would require 100 CatalogProductUpdateAction's. The purpose of `CatalogProductSetStatusAction`is to conveniently batch those related updated into one transaction.

As for the enforcement of the `status` field, the smart contract could throw an error if the catalog_product does not have a `status` field in its additional fields. I want to avoid creating a `catalog_product` message just for enforcing one field. The reusing of Grid Product would get us the functionality we need, without the overhead of managing an additional message type that is only slightly different from the base.

Reviewer 9: Ok. I added a PR to your branch expressing that alternative.

    enum Status {
        INACTIVE = 0;
        ACTIVE = 1;
        DISCONTINUED = 2;
    }
    // catalog_id and product_id are used in deriving the state address
    repeated string catalog_ids = 1;
    string catalog_product_id = 2;
    Status catalog_product_status  = 4;
    // Reason for the change

>     string reason = 5;

Reviewer 9: What happens to this reason during execution? I don't see a place where it gets stored in state.

Author: Oops added that field after writing up the schema. 
The `reason` field would need to be defined in the catalog_product schema like so:

```
        PropertyDefinition(
            name="status_change_reason",
            data_type=PropertyDefinition.DataType.STRING,
            description="The reason for the status change",
            required=True
        ),
```

An implementation in catalog_product would look like:
```
        PropertyValue(
            name="status_change_reason",
            data_type=PropertyDefinition.DataType.STRING,
            string_value=""
        ),
```

`reason` would probably be best to keep as an empty string upon initialization until the status of the catalog_product is actually changed.

Also, I've updated that field name to be synonymous with the catalog_product schema change above.
```
    string catalog_product_id = 2;
    Status catalog_product_status  = 4;
    // Reason for the change
    `string status_change_reason = 5;`
```

}
```

Validation requirements:

- The catalog_product must be able to be updated or the transaction is invalid 
(i.e. meaning the status is not "DISCONTINUED")

> - If any catalog_product with product_id does not exist the transaction is

Reviewer: s/does not exist the/does not exist, the/

Author: ✅

- Feature Name: Product Catalog
- Start Date: 2019-07-12
- RFC PR:
- Hyperledger Grid Issue: [GRID-95](https://jira.hyperledger.org/browse/GRID-95)

# Summary

[summary]: #summary

This RFC proposes a generic implementation for a Hyperledger Grid _Product
Catalog_. Product Catalogs will be represented by a unique identifier, which

> will

Author: ✅

be referenced by the _catalog_products_ included in the catalog. That
catalog_id
can be used to share a grouping of products that can be shared with one or more
organizations. The Grid Catalog will contain 4 fields. A **catalog_id**, a
**catalog_owner**, a **Name** for the catalog, and a repeated field of
PropertyValues for custom fields.

In addition, a base catalog product schema is included as a way to demonstrate
enforcement of additional product properties. Any grid product added to a
catalog

> will adhere to the "catalog*product" schema (or another custom-defined schema).

Author: ✅

are supported:

**catalog actions:**

- CatalogCreate - create a catalog and store it in state
- CatalogUpdate - update a catalog in state
- \*CatalogDelete - deletes a catalog from state and associated catalog_products

**catalog_product actions:**

- CatalogProductCreate - create a catalog_product and stores it in state
- CatalogProductUpdate - updates a catalog_product in state
- CatalogProductDelete - delete a catalog_product from state

**catalog_product operation:**

- \*CatalogProductSetStatus - changes the "status" of a catalog_product to an
  active,
  inactive, or discontinued state

> The catalog_product actions/operation will accept a parameter list of the

Author: ✅

representation.

All Grid addresses are prefixed by the 6-hex-character namespace prefix
"621dee", Catalogs are further prefixed under the Grid namespace with reserved
enumerations of "03" ("00", "01", and "02" being reserved for other purposes)
indicating "Grid Catalogs" and an additional org_id. Currently in Grid Product,
org_ids are a [6-10 digit GS1 company
prefix](https://www.gs1-us.info/gs1-company-prefix/).
To keep the prefix length consistent, any GS1 company prefix less than 10
digits will
be left padded with the hexadecimal representation of the number 10 ("a") until
it is
10 digits in length. This composite address will enable inexpensive catalog
operations.

Therefore, all addresses starting with:

```
"621dee" + "03" + "aaaa123456" // org_id (in this case a gs1_company_prefix)
```

are Grid GS1 Catalogs identified by the hash of the catalog name which can be
referenced by a "catalog_product" to group products to a specific catalog.

The catalog_id format consists of 70-digit "alphanumeric string" which include
a fixed amount of internal "0" padding. After the 18-hex-characters that are

> consumed by the grid namespace prefix, the catalog, and gs1_company_prefix

Reviewer 7: "the catalog" should say the "catalog prefix"

Author: ✅

there

> are 52 hex characters remaining in the address. The 18 digits of the

Author: ✅

catalog_id can

> be right padded with 35-hex-character zeroes and right padded with

Reviewer 7: Should this say left padded? (the first right padded statement)

Author: ✅

The 12 to 14 digits of the GTIN can be left padded with 40 to 42-hex-character
zeroes and right padded with 2-hex-character zeroes to accommodate potential
future storage associated with the GS1 Product representation, for example:

```
"621dee" + "02" + "01" + "cad4e781e797712" + "00000000000000000000000000000"
+ 14-character "numeric string" product_id + "00" // product_id == GTIN
```

A full Catalog Product address using the example GTIN from
https://www.gtin.info/ would therefore be:

```
"621dee0201cad4e781e797712000000000000000000000000000000001234560001200"
```

### Catalog Product Schema Definition

This schema defines the additional attributes required of a grid product, such
that
it may be considered a "catalog product". Catalog products should be grouped
with
other catalog products. Once grouped, the grid catalog can be shared with other
participants in a grid network.

Any grid product added to a catalog (thus becoming a catalog*product) will need
to
contain a reference fields to the original product (product_id,
product_namespace,
and owner) and a catalog_id. A catalog product schema may also be defined to

> required additional fields for that catalog_product. This example schema

Reviewer 7: required -> require

Author: ✅

    ])
```

## Transaction Payload and Execution

> The following payloads contain an action enum and the associated action

Reviewer 7: payloads -> payload

Author: ✅

payload.
This allows for the action payload to be dispatched to the appropriate logic.
Only the defined actions (enum) are available, and only one Action should be
performed in a payload.

### CatalogPayload Transaction

```
message CatalogPayload {
    enum Actions {
        UNSET_ACTION = 0;
        CATALOG_CREATE = 1;
        CATALOG_UPDATE = 2;
        CATALOG_DELETE = 3;

>         CATALOG_PRODUCT_CREATE = 4;

Reviewer 7: This should start at 100 to leave room for more Catalog actions

Author: ✅

        CATALOG_PRODUCT_UPDATE = 5;
        CATALOG_PRODUCT_DELETE = 6;

>          SET_CATALOG_PRODUCT_STATUS = 7;

Reviewer 7: This should be CATALOG_PRODUCT_SET_STATUS

Author: ✅

Address of the Catalog created

### CatalogUpdateAction

CatalogUpdateAction adds a new catalog to state. The transaction should be
submitted by
an agent, which is identified by its signing key, acting on behalf of the
organization
that corresponds to the owner in the update transaction. (Organizations and
agents are
defined by the Pike smart contract.)

```
message CatalogUpdateAction {
    // GS1 Company Prefix from owner + catalog_id are use as
    // a composite key for determining the state address
    string owner = 1;
    string catalog_id = 2;
    string name = 3;
    repeated PropertyValues properties = 4;
}
```

Validation requirements:

> - If a catalog with catalog_id exists the transaction is invalid.

Reviewer 7: Why was this deleted

Reviewer 7: (same in other locations)

Author: ✅

## Catalog_Product Actions

### CatalogProductCreateAction

> The CatalogProductCreateAction adds a new catalog_product to state. The catalog_product references a Grid Product for the shared item level master data. The

Reviewer 7: line too long

Author: ✅

    string catalog_id = 1;
    string product_id = 2;
    // this will replace all properties currently defined
    repeated PropertyValues properties = 4;
}
```

Validation requirements:

- If a catalog_product with catalog_product_id does not exist the transaction
  is invalid.
- The signer of the transaction must be an agent in the Pike state and must
  belong to an
  organization in Pike state, otherwise the transaction is invalid.
  The owner in the product must match the organization that the agent belongs to,
  otherwise the transaction is invalid.
- The agent must have the permission can_update_product for the organization,
  otherwise
  the transaction is invalid.
- The properties must be valid for the catalog_product schema; its properties
  must only contain properties that are included in the catalog_product Schema.

The inputs for CatalogProductUpdateAction must include:

- Address of the Agent submitting the transaction
- Address of the Organization the Product is being updated for
- Address of the Product to be updated

The outputs for CatalogProductUpdateAction must include:

- Address of the updated catalog_product

### CatalogProductDeleteAction

> CatalogProductDeleteAction removes an existing product from state. The

Reviewer 7: This remove the catalog product, not the grid product correct? Change "existing product" to "existing catalog product"

Author: ✅

transaction should be
submitted by an agent, identified by its signing key, acting on behalf of the
organization
that corresponds to the org_id in the product being deleted. (Organizations and
agents are
defined by the Pike smart contract.)

```
message CatalogProductDeleteAction {
    // catalog_id and product_id are used in deriving the state address
    string catalog_id = 1;
    string product_id = 2;
}
```

If the Grid setting grid.product.allow_delete is set to false, this transaction
is invalid. (sys admin setting)
The default value for grid.product.allow_delete is true. This setting is stored
using the
Sawtooth Settings smart contract, more information can be found here.

Validation requirements:

> - If a catalog_product with product_id does not exist the transaction is

Reviewer 7: everywhere you say product_id meaning the id for the catalog product it should say catalog_product_id

Author: ✅

CatalogProductSetStatusAction updates an existing catalog_product in state. The
transaction should be submitted by an agent, identified by its signing key, 
acting on behalf of an organization that corresponds to the owner in the 
product being updated. (Organizations and agents are defined by the Pike smart 
contract.)

> The implementation of the CatalogProductSetStatusAction will handling iterating 

Reviewer 7: will handling -> handle

Author: ✅

> through the list of catalog_ids, and updating the status of the 

Reviewer 7: updating -> update

Author: ✅

Validation requirements:

> - The catalog_product must be able to be updated or the transaction is invalid

Reviewer 7: I'm not sure what this means. This should be enough information for someone to implement the smart contract.

Author: added more info 
> (i.e. meaning the status is not "DISCONTINUED")

> - If a catalog_product with product_id does not exist the transaction is

Reviewer 7: This should clarify that if ANY of the catalog_products referenced in the transaction do not exist the transaction is invalid.

Author: ✅
